{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7714548c-2b29-4c6e-8d32-3c9274280789",
   "metadata": {},
   "source": [
    "## CDM ONTOLOGIES WORKFLOW"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad1b0498-54cd-480c-9d7e-96bfe6ce4141",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Full workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44ffde61-3391-4db7-a1bf-bdd254974c10",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting CDM Ontologies Workflow...\n",
      "\n",
      "1. Analyzing Core Ontologies...\n",
      "\n",
      "Analysis Results:\n",
      "Skipping download, envo.owl already present\n",
      "\n",
      "File: envo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/envo.owl\n",
      "  Own terms: 4385\n",
      "  External terms: 3043\n",
      "  Classification: Non-Base. Base version available in OBO Foundry: http://purl.obolibrary.org/obo/envo/envo-base.owl.\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 2984\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000001\n",
      "    http://purl.obolibrary.org/obo/BFO_0000002\n",
      "    http://purl.obolibrary.org/obo/BFO_0000003\n",
      "    http://purl.obolibrary.org/obo/BFO_0000004\n",
      "    http://purl.obolibrary.org/obo/BFO_0000006\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000000\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000001\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000002\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000003\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000004\n",
      "  First 5 external terms:\n",
      "    http://geneontology.org/foo/applies-pattern\n",
      "    http://purl.obolibrary.org/obo/BFO_0000001\n",
      "    http://purl.obolibrary.org/obo/BFO_0000002\n",
      "    http://purl.obolibrary.org/obo/BFO_0000003\n",
      "    http://purl.obolibrary.org/obo/BFO_0000004\n",
      "Skipping download, go.owl already present\n",
      "\n",
      "File: go.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/go.owl\n",
      "  Own terms: 51567\n",
      "  External terms: 73\n",
      "  Classification: Non-Base. Base version available in OBO Foundry: http://purl.obolibrary.org/obo/go/go-base.owl.\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 53\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000066\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000233\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/GO_0000001\n",
      "    http://purl.obolibrary.org/obo/GO_0000002\n",
      "    http://purl.obolibrary.org/obo/GO_0000003\n",
      "    http://purl.obolibrary.org/obo/GO_0000004\n",
      "    http://purl.obolibrary.org/obo/GO_0000005\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000066\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000116\n",
      "Skipping download, ncbitaxon.owl already present\n",
      "\n",
      "File: ncbitaxon.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/ncbitaxon.owl\n",
      "  Own terms: 2594866\n",
      "  External terms: 31\n",
      "  Classification: Potential Base (missing -base hint in .owl file name). No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 28\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/OMO_0003003\n",
      "    http://purl.obolibrary.org/obo/OMO_0003006\n",
      "    http://purl.obolibrary.org/obo/OMO_0003007\n",
      "    http://purl.obolibrary.org/obo/OMO_0003012\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon#_species_group\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon#_species_subgroup\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon#_taxonomic_rank\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon_1\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon_10\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/OMO_0003003\n",
      "    http://purl.obolibrary.org/obo/OMO_0003006\n",
      "    http://purl.obolibrary.org/obo/OMO_0003007\n",
      "    http://purl.obolibrary.org/obo/OMO_0003012\n",
      "Skipping download, chebi.owl already present\n",
      "\n",
      "File: chebi.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/chebi.owl\n",
      "  Own terms: 220224\n",
      "  External terms: 48\n",
      "  Classification: Potential Base (missing -base hint in .owl file name). No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 29\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/RO_0000087\n",
      "    http://purl.obolibrary.org/obo/chebi#1_STAR\n",
      "    http://purl.obolibrary.org/obo/chebi#2_STAR\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/CHEBI_1\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10\n",
      "    http://purl.obolibrary.org/obo/CHEBI_100\n",
      "    http://purl.obolibrary.org/obo/CHEBI_1000\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10000\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000231\n",
      "    http://purl.obolibrary.org/obo/IAO_0100001\n",
      "    http://purl.obolibrary.org/obo/RO_0000087\n",
      "\n",
      "Analysis complete!\n",
      "\n",
      "2. Analyzing Non-Core Ontologies...\n",
      "\n",
      "Processing external terms from core ontologies...\n",
      "Downloading regular version of po...\n",
      "Successfully downloaded: po.owl\n",
      "Downloading base version of caro...\n",
      "Successfully downloaded: caro-base.owl\n",
      "Downloading base version of fao...\n",
      "Successfully downloaded: fao-base.owl\n",
      "Downloading base version of uberon...\n",
      "Successfully downloaded: uberon-base.owl\n",
      "Downloading base version of cl...\n",
      "Successfully downloaded: cl-base.owl\n",
      "Downloading base version of obi...\n",
      "Successfully downloaded: obi-base.owl\n",
      "Downloading base version of pato...\n",
      "Successfully downloaded: pato-base.owl\n",
      "Downloading base version of ro...\n",
      "Successfully downloaded: ro-base.owl\n",
      "Downloading regular version of bfo...\n",
      "Successfully downloaded: bfo.owl\n",
      "Downloading regular version of iao...\n",
      "Successfully downloaded: iao.owl\n",
      "Downloading base version of pco...\n",
      "Successfully downloaded: pco-base.owl\n",
      "Downloading regular version of foodon...\n",
      "Successfully downloaded: foodon.owl\n",
      "Downloading regular version of omo...\n",
      "Successfully downloaded: omo.owl\n",
      "\n",
      "Updating ontologies.txt...\n",
      "\n",
      "Processing Additional OBO Foundry, PyOBO and In-house ontologies...\n",
      "Downloading uo.owl...\n",
      "Successfully downloaded: uo.owl\n",
      "Downloading eccode.owl.gz...\n",
      "Successfully downloaded and decompressed: eccode.owl\n",
      "Downloading pfam.owl...\n",
      "Successfully downloaded: pfam.owl\n",
      "Downloading rhea.owl.gz...\n",
      "Successfully downloaded and decompressed: rhea.owl\n",
      "\n",
      "Analyzing non-core ontologies...\n",
      "\n",
      "Analyzing non-core ontologies:\n",
      "\n",
      "File: bfo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/bfo.owl\n",
      "  Own terms: 37\n",
      "  External terms: 21\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 14\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000111\n",
      "    http://purl.obolibrary.org/obo/IAO_0000112\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000116\n",
      "    http://purl.obolibrary.org/obo/IAO_0000117\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000001\n",
      "    http://purl.obolibrary.org/obo/BFO_0000002\n",
      "    http://purl.obolibrary.org/obo/BFO_0000003\n",
      "    http://purl.obolibrary.org/obo/BFO_0000004\n",
      "    http://purl.obolibrary.org/obo/BFO_0000006\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000111\n",
      "    http://purl.obolibrary.org/obo/IAO_0000112\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000116\n",
      "    http://purl.obolibrary.org/obo/IAO_0000117\n",
      "\n",
      "File: eccode.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/eccode.owl\n",
      "  Own terms: 0\n",
      "  External terms: 249359\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 8743\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0100001\n",
      "    http://purl.obolibrary.org/obo/RO_0002327\n",
      "    http://purl.obolibrary.org/obo/RO_0002351\n",
      "    http://purl.obolibrary.org/obo/eccode.owl\n",
      "    http://www.geneontology.org/formats/oboInOwl#hasExactSynonym\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/GO_0000010\n",
      "    http://purl.obolibrary.org/obo/GO_0000016\n",
      "    http://purl.obolibrary.org/obo/GO_0000034\n",
      "    http://purl.obolibrary.org/obo/GO_0000048\n",
      "    http://purl.obolibrary.org/obo/GO_0000121\n",
      "\n",
      "File: foodon.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/foodon.owl\n",
      "  Own terms: 28193\n",
      "  External terms: 10746\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 10699\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000025\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000322\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000360\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000363\n",
      "    http://purl.obolibrary.org/obo/AGRO_00002071\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000000\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000015\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000070\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000071\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000072\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000025\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000322\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000360\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000363\n",
      "    http://purl.obolibrary.org/obo/AGRO_00002071\n",
      "\n",
      "File: iao.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/iao.owl\n",
      "  Own terms: 283\n",
      "  External terms: 130\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 101\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000008\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000032\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000033\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000522\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000524\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000001\n",
      "    http://purl.obolibrary.org/obo/IAO_0000002\n",
      "    http://purl.obolibrary.org/obo/IAO_0000003\n",
      "    http://purl.obolibrary.org/obo/IAO_0000004\n",
      "    http://purl.obolibrary.org/obo/IAO_0000005\n",
      "  First 5 external terms:\n",
      "    http://protege.stanford.edu/plugins/owl/protege#defaultLanguage\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000008\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000032\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000033\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000522\n",
      "\n",
      "File: omo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/omo.owl\n",
      "  Own terms: 18\n",
      "  External terms: 107\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 84\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000002\n",
      "    http://purl.obolibrary.org/obo/IAO_0000027\n",
      "    http://purl.obolibrary.org/obo/IAO_0000030\n",
      "    http://purl.obolibrary.org/obo/IAO_0000078\n",
      "    http://purl.obolibrary.org/obo/IAO_0000102\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/OMO_0001000\n",
      "    http://purl.obolibrary.org/obo/OMO_0001001\n",
      "    http://purl.obolibrary.org/obo/OMO_0001002\n",
      "    http://purl.obolibrary.org/obo/OMO_0002000\n",
      "    http://purl.obolibrary.org/obo/OMO_0003000\n",
      "  First 5 external terms:\n",
      "    http://protege.stanford.edu/plugins/owl/protege#defaultLanguage\n",
      "    http://purl.obolibrary.org/obo/IAO_0000002\n",
      "    http://purl.obolibrary.org/obo/IAO_0000027\n",
      "    http://purl.obolibrary.org/obo/IAO_0000030\n",
      "    http://purl.obolibrary.org/obo/IAO_0000078\n",
      "\n",
      "File: pfam.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/pfam.owl\n",
      "  Own terms: 0\n",
      "  External terms: 24553\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 23798\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/pfam.owl\n",
      "    http://purl.obolibrary.org/obo/pfam_PF00001\n",
      "    http://purl.obolibrary.org/obo/pfam_PF00002\n",
      "    http://purl.obolibrary.org/obo/pfam_PF00003\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/pfam.clan_CL0001\n",
      "    http://purl.obolibrary.org/obo/pfam.clan_CL0003\n",
      "    http://purl.obolibrary.org/obo/pfam.clan_CL0004\n",
      "    http://purl.obolibrary.org/obo/pfam.clan_CL0005\n",
      "\n",
      "File: po.owl\n",
      "  Has imports: Yes\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/po.owl\n",
      "  Own terms: 1996\n",
      "  External terms: 67\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 50\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000062\n",
      "    http://purl.obolibrary.org/obo/BFO_0000063\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/PO_0000001\n",
      "    http://purl.obolibrary.org/obo/PO_0000002\n",
      "    http://purl.obolibrary.org/obo/PO_0000003\n",
      "    http://purl.obolibrary.org/obo/PO_0000004\n",
      "    http://purl.obolibrary.org/obo/PO_0000005\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000062\n",
      "    http://purl.obolibrary.org/obo/BFO_0000063\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "\n",
      "File: rhea.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/rhea.owl\n",
      "  Own terms: 0\n",
      "  External terms: 323609\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 67894\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/RO_0000056\n",
      "    http://purl.obolibrary.org/obo/RO_0000057\n",
      "    http://purl.obolibrary.org/obo/RO_0002233\n",
      "    http://purl.obolibrary.org/obo/RO_0002234\n",
      "    http://purl.obolibrary.org/obo/RO_0002327\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/CHEBI_100\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10015\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10023\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10036\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10106\n",
      "\n",
      "File: uo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/uo.owl\n",
      "  Own terms: 572\n",
      "  External terms: 5\n",
      "  Classification: Potential Base (missing -base hint in .owl file name). No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 1\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/uo.owl\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/UO_0000000\n",
      "    http://purl.obolibrary.org/obo/UO_0000001\n",
      "    http://purl.obolibrary.org/obo/UO_0000002\n",
      "    http://purl.obolibrary.org/obo/UO_0000003\n",
      "    http://purl.obolibrary.org/obo/UO_0000004\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/has_prefix\n",
      "    http://purl.obolibrary.org/obo/is_unit_of\n",
      "    http://purl.obolibrary.org/obo/uo.owl\n",
      "    http://www.geneontology.org/formats/oboInOwl#hasExactSynonym\n",
      "    http://www.geneontology.org/formats/oboInOwl#hasRelatedSynonym\n",
      "\n",
      "Analysis complete!\n",
      "\n",
      "3. Creating Pseudo Base Ontologies...\n",
      "Using ROBOT at: /cdm_shared_workspace/user_shared_workspace/Jose/install_stuff/robot/robot\n",
      "Processing uo.owl...\n",
      "Executing command:\n",
      "robot remove --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/uo.owl --base-iri http://purl.obolibrary.org/obo/HAS_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl\n",
      "Created base version for uo.owl: uo-base.owl\n",
      "Processing omo.owl...\n",
      "Executing command:\n",
      "robot remove --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/omo.owl --base-iri http://purl.obolibrary.org/obo/IAO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl\n",
      "Created base version for omo.owl: omo-base.owl\n",
      "Processing iao.owl...\n",
      "Executing command:\n",
      "robot remove --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/iao.owl --base-iri http://purl.obolibrary.org/obo/BFO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl\n",
      "Created base version for iao.owl: iao-base.owl\n",
      "Processing bfo.owl...\n",
      "Executing command:\n",
      "robot remove --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/bfo.owl --base-iri http://purl.obolibrary.org/obo/BFO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl\n",
      "Created base version for bfo.owl: bfo-base.owl\n",
      "Processing foodon.owl...\n",
      "Executing command:\n",
      "robot remove --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/foodon.owl --base-iri http://purl.obolibrary.org/obo/COB_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl\n",
      "Created base version for foodon.owl: foodon-base.owl\n",
      "Copying PyOBO ontology to main directory: eccode.owl\n",
      "Processing po.owl...\n",
      "Executing command:\n",
      "robot remove --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/po.owl --base-iri http://purl.obolibrary.org/obo/IAO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl\n",
      "Created base version for po.owl: po-base.owl\n",
      "Copying PyOBO ontology to main directory: rhea.owl\n",
      "Copying PyOBO ontology to main directory: pfam.owl\n",
      "\n",
      "Processing complete!\n",
      "\n",
      "4. Merging Ontologies...\n",
      "Using ROBOT at: /cdm_shared_workspace/user_shared_workspace/Jose/install_stuff/robot/robot\n",
      "Looking for ontology files in: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl\n",
      "Found 25 ontology files:\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/envo.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl\n",
      "Created list of merged ontologies at: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/CDM_merged_ontologies.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/envo.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/go.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/CDM_merged_ontologies.owl\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "# Import all required functions\n",
    "from analyze_core_ontologies import analyze_core_ontologies\n",
    "import analyze_non_core_ontologies\n",
    "from create_pseudo_base_ontology import create_pseudo_base_ontologies\n",
    "from merge_ontologies import merge_ontologies\n",
    "from create_semantic_sql_db import create_semantic_sql_db\n",
    "from extract_sql_tables_to_tsv import extract_sql_tables_to_tsv\n",
    "\n",
    "# Print workflow start\n",
    "print(\"Starting CDM Ontologies Workflow...\")\n",
    "\n",
    "print(\"\\n1. Analyzing Core Ontologies...\")\n",
    "analyze_core_ontologies(repo_path)\n",
    "\n",
    "print(\"\\n2. Analyzing Non-Core Ontologies...\")\n",
    "analyze_non_core_ontologies.analyze_non_core_ontologies(repo_path)\n",
    "\n",
    "print(\"\\n3. Creating Pseudo Base Ontologies...\")\n",
    "create_pseudo_base_ontologies(repo_path)\n",
    "\n",
    "print(\"\\n4. Merging Ontologies...\")\n",
    "if not merge_ontologies(repo_path):\n",
    "    raise Exception(\"Ontology merge failed\")\n",
    "\n",
    "print(\"\\n5. Creating Semantic SQL Database...\")\n",
    "if not create_semantic_sql_db(repo_path):\n",
    "    raise Exception(\"Database creation failed\")\n",
    "\n",
    "print(\"\\n6. Extracting SQL Tables to TSV...\")\n",
    "if not extract_sql_tables_to_tsv(repo_path):\n",
    "    raise Exception(\"TSV extraction failed\")\n",
    "\n",
    "print(\"\\nWorkflow completed successfully!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12e608d7-99fa-4d00-8cdd-235e0588112f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Worflow in seperate steps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e31d686-e69a-4085-8888-08545ee08eeb",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Analyze and Download Core Ontologies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "210f3f80-b954-4b2c-a933-906b562a02dd",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Analysis Results:\n",
      "Downloading envo.owl...\n",
      "Successfully downloaded: envo.owl\n",
      "\n",
      "File: envo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/envo.owl\n",
      "  Own terms: 4514\n",
      "  External terms: 2858\n",
      "  Classification: Non-Base. Base version available in OBO Foundry: http://purl.obolibrary.org/obo/envo/envo-base.owl.\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 2807\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000001\n",
      "    http://purl.obolibrary.org/obo/BFO_0000002\n",
      "    http://purl.obolibrary.org/obo/BFO_0000003\n",
      "    http://purl.obolibrary.org/obo/BFO_0000004\n",
      "    http://purl.obolibrary.org/obo/BFO_0000006\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000000\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000001\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000002\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000003\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000004\n",
      "  First 5 external terms:\n",
      "    http://geneontology.org/foo/applies-pattern\n",
      "    http://purl.obolibrary.org/obo/BFO_0000001\n",
      "    http://purl.obolibrary.org/obo/BFO_0000002\n",
      "    http://purl.obolibrary.org/obo/BFO_0000003\n",
      "    http://purl.obolibrary.org/obo/BFO_0000004\n",
      "Downloading go.owl...\n",
      "Successfully downloaded: go.owl\n",
      "\n",
      "File: go.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/go.owl\n",
      "  Own terms: 51567\n",
      "  External terms: 73\n",
      "  Classification: Non-Base. Base version available in OBO Foundry: http://purl.obolibrary.org/obo/go/go-base.owl.\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 53\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000066\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000233\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/GO_0000001\n",
      "    http://purl.obolibrary.org/obo/GO_0000002\n",
      "    http://purl.obolibrary.org/obo/GO_0000003\n",
      "    http://purl.obolibrary.org/obo/GO_0000004\n",
      "    http://purl.obolibrary.org/obo/GO_0000005\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000066\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000116\n",
      "Downloading ncbitaxon.owl...\n",
      "Successfully downloaded: ncbitaxon.owl\n",
      "\n",
      "File: ncbitaxon.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/ncbitaxon.owl\n",
      "  Own terms: 2613937\n",
      "  External terms: 31\n",
      "  Classification: Potential Base (missing -base hint in .owl file name). No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 28\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/OMO_0003003\n",
      "    http://purl.obolibrary.org/obo/OMO_0003006\n",
      "    http://purl.obolibrary.org/obo/OMO_0003007\n",
      "    http://purl.obolibrary.org/obo/OMO_0003012\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon#_species_group\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon#_species_subgroup\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon#_taxonomic_rank\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon_1\n",
      "    http://purl.obolibrary.org/obo/NCBITaxon_10\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/OMO_0003003\n",
      "    http://purl.obolibrary.org/obo/OMO_0003006\n",
      "    http://purl.obolibrary.org/obo/OMO_0003007\n",
      "    http://purl.obolibrary.org/obo/OMO_0003012\n",
      "Downloading chebi.owl...\n",
      "Successfully downloaded: chebi.owl\n",
      "\n",
      "File: chebi.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/chebi.owl\n",
      "  Own terms: 220816\n",
      "  External terms: 48\n",
      "  Classification: Potential Base (missing -base hint in .owl file name). No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 29\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/RO_0000087\n",
      "    http://purl.obolibrary.org/obo/chebi#1_STAR\n",
      "    http://purl.obolibrary.org/obo/chebi#2_STAR\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/CHEBI_1\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10\n",
      "    http://purl.obolibrary.org/obo/CHEBI_100\n",
      "    http://purl.obolibrary.org/obo/CHEBI_1000\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10000\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000231\n",
      "    http://purl.obolibrary.org/obo/IAO_0100001\n",
      "    http://purl.obolibrary.org/obo/RO_0000087\n",
      "\n",
      "Analysis complete!\n",
      "CPU times: user 3min 14s, sys: 29.7 s, total: 3min 44s\n",
      "Wall time: 8min 11s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the analysis\n",
    "from analyze_core_ontologies import analyze_core_ontologies\n",
    "\n",
    "# Run the analysis\n",
    "analyze_core_ontologies(repo_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5810db4-bf44-4905-8de0-13156df573b9",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Analyze and Download Non-Core Ontologies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d5b35a0f-53f8-4b85-9149-2ba2d9a0efde",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing external terms from core ontologies...\n",
      "Downloading base version of ro...\n",
      "Successfully downloaded: ro-base.owl\n",
      "Downloading base version of pato...\n",
      "Successfully downloaded: pato-base.owl\n",
      "Downloading base version of obi...\n",
      "Successfully downloaded: obi-base.owl\n",
      "Downloading base version of fao...\n",
      "Successfully downloaded: fao-base.owl\n",
      "Downloading regular version of po...\n",
      "Successfully downloaded: po.owl\n",
      "Downloading base version of pco...\n",
      "Successfully downloaded: pco-base.owl\n",
      "Downloading regular version of iao...\n",
      "Successfully downloaded: iao.owl\n",
      "Downloading base version of uberon...\n",
      "Successfully downloaded: uberon-base.owl\n",
      "Downloading regular version of omo...\n",
      "Successfully downloaded: omo.owl\n",
      "Downloading regular version of bfo...\n",
      "Successfully downloaded: bfo.owl\n",
      "Downloading regular version of foodon...\n",
      "Successfully downloaded: foodon.owl\n",
      "\n",
      "Updating ontologies.txt...\n",
      "\n",
      "Processing Additional OBO Foundry, PyOBO and In-house ontologies...\n",
      "Downloading uo.owl...\n",
      "Successfully downloaded: uo.owl\n",
      "Downloading taxrank.owl...\n",
      "Successfully downloaded: taxrank.owl\n",
      "Downloading so.owl...\n",
      "Successfully downloaded: so.owl\n",
      "Downloading gtdb.owl...\n",
      "Successfully downloaded: gtdb.owl\n",
      "Downloading eccode.owl.gz...\n",
      "Successfully downloaded and decompressed: eccode.owl\n",
      "Downloading pfam.owl...\n",
      "Successfully downloaded: pfam.owl\n",
      "Downloading rhea.owl.gz...\n",
      "Successfully downloaded and decompressed: rhea.owl\n",
      "Downloading credit.owl...\n",
      "Successfully downloaded: credit.owl\n",
      "Downloading ror.owl.gz...\n",
      "Successfully downloaded and decompressed: ror.owl\n",
      "Downloading interpro.owl...\n",
      "Successfully downloaded: interpro.owl\n",
      "\n",
      "Analyzing non-core ontologies...\n",
      "\n",
      "Analyzing non-core ontologies:\n",
      "\n",
      "File: bfo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/bfo.owl\n",
      "  Own terms: 37\n",
      "  External terms: 21\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 14\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000111\n",
      "    http://purl.obolibrary.org/obo/IAO_0000112\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000116\n",
      "    http://purl.obolibrary.org/obo/IAO_0000117\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000001\n",
      "    http://purl.obolibrary.org/obo/BFO_0000002\n",
      "    http://purl.obolibrary.org/obo/BFO_0000003\n",
      "    http://purl.obolibrary.org/obo/BFO_0000004\n",
      "    http://purl.obolibrary.org/obo/BFO_0000006\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000111\n",
      "    http://purl.obolibrary.org/obo/IAO_0000112\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000116\n",
      "    http://purl.obolibrary.org/obo/IAO_0000117\n",
      "\n",
      "File: credit.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: https://w3id.org/biopragmatics/resources/credit/credit.ofn\n",
      "  Own terms: 0\n",
      "  External terms: 19\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 15\n",
      "  First 5 external terms that are subject of triples:\n",
      "    https://credit.niso.org/contributor-roles/conceptualization\n",
      "    https://credit.niso.org/contributor-roles/data-curation\n",
      "    https://credit.niso.org/contributor-roles/formal-analysis\n",
      "    https://credit.niso.org/contributor-roles/funding-acquisition\n",
      "    https://credit.niso.org/contributor-roles/investigation\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.org/dc/terms/description\n",
      "    http://purl.org/dc/terms/license\n",
      "    http://purl.org/dc/terms/title\n",
      "    http://www.geneontology.org/formats/oboInOwl#auto-generated-by\n",
      "    https://credit.niso.org/contributor-roles/conceptualization\n",
      "\n",
      "File: eccode.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/eccode.owl\n",
      "  Own terms: 0\n",
      "  External terms: 249359\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 8743\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0100001\n",
      "    http://purl.obolibrary.org/obo/RO_0002327\n",
      "    http://purl.obolibrary.org/obo/RO_0002351\n",
      "    http://purl.obolibrary.org/obo/eccode.owl\n",
      "    http://www.geneontology.org/formats/oboInOwl#hasExactSynonym\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/GO_0000010\n",
      "    http://purl.obolibrary.org/obo/GO_0000016\n",
      "    http://purl.obolibrary.org/obo/GO_0000034\n",
      "    http://purl.obolibrary.org/obo/GO_0000048\n",
      "    http://purl.obolibrary.org/obo/GO_0000121\n",
      "\n",
      "File: foodon.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/foodon.owl\n",
      "  Own terms: 28805\n",
      "  External terms: 10966\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 10934\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000025\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000322\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000360\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000363\n",
      "    http://purl.obolibrary.org/obo/AGRO_00002071\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000000\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000015\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000070\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000071\n",
      "    http://purl.obolibrary.org/obo/FOODON_00000072\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000025\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000322\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000360\n",
      "    http://purl.obolibrary.org/obo/AGRO_00000363\n",
      "    http://purl.obolibrary.org/obo/AGRO_00002071\n",
      "\n",
      "File: gtdb.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: https://w3id.org/biopragmatics/resources/gtdb/gtdb.ofn\n",
      "  Own terms: 0\n",
      "  External terms: 146394\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 146389\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000700\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000001\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000002\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000003\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000004\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000700\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000001\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000002\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000003\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000004\n",
      "\n",
      "File: iao.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/iao.owl\n",
      "  Own terms: 283\n",
      "  External terms: 130\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 101\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000008\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000032\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000033\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000522\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000524\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000001\n",
      "    http://purl.obolibrary.org/obo/IAO_0000002\n",
      "    http://purl.obolibrary.org/obo/IAO_0000003\n",
      "    http://purl.obolibrary.org/obo/IAO_0000004\n",
      "    http://purl.obolibrary.org/obo/IAO_0000005\n",
      "  First 5 external terms:\n",
      "    http://protege.stanford.edu/plugins/owl/protege#defaultLanguage\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000008\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000032\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000033\n",
      "    http://purl.obolibrary.org/obo/APOLLO_SV_00000522\n",
      "\n",
      "File: interpro.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: https://w3id.org/biopragmatics/resources/interpro/interpro.ofn\n",
      "  Own terms: 0\n",
      "  External terms: 52757\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 47053\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IPR_IPR000001\n",
      "    http://purl.obolibrary.org/obo/IPR_IPR000003\n",
      "    http://purl.obolibrary.org/obo/IPR_IPR000006\n",
      "    http://purl.obolibrary.org/obo/IPR_IPR000007\n",
      "    http://purl.obolibrary.org/obo/IPR_IPR000008\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/GO_0000001\n",
      "    http://purl.obolibrary.org/obo/GO_0000002\n",
      "    http://purl.obolibrary.org/obo/GO_0000009\n",
      "    http://purl.obolibrary.org/obo/GO_0000011\n",
      "    http://purl.obolibrary.org/obo/GO_0000012\n",
      "\n",
      "File: omo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/omo.owl\n",
      "  Own terms: 18\n",
      "  External terms: 107\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 84\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000002\n",
      "    http://purl.obolibrary.org/obo/IAO_0000027\n",
      "    http://purl.obolibrary.org/obo/IAO_0000030\n",
      "    http://purl.obolibrary.org/obo/IAO_0000078\n",
      "    http://purl.obolibrary.org/obo/IAO_0000102\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/OMO_0001000\n",
      "    http://purl.obolibrary.org/obo/OMO_0001001\n",
      "    http://purl.obolibrary.org/obo/OMO_0001002\n",
      "    http://purl.obolibrary.org/obo/OMO_0002000\n",
      "    http://purl.obolibrary.org/obo/OMO_0003000\n",
      "  First 5 external terms:\n",
      "    http://protege.stanford.edu/plugins/owl/protege#defaultLanguage\n",
      "    http://purl.obolibrary.org/obo/IAO_0000002\n",
      "    http://purl.obolibrary.org/obo/IAO_0000027\n",
      "    http://purl.obolibrary.org/obo/IAO_0000030\n",
      "    http://purl.obolibrary.org/obo/IAO_0000078\n",
      "\n",
      "File: pfam.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: https://w3id.org/biopragmatics/resources/pfam/pfam.ofn\n",
      "  Own terms: 0\n",
      "  External terms: 24850\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 24077\n",
      "  First 5 external terms that are subject of triples:\n",
      "    https://w3id.org/biopragmatics/resources/pfam/pfam.ofn\n",
      "    https://www.ebi.ac.uk/interpro/entry/pfam/PF00001\n",
      "    https://www.ebi.ac.uk/interpro/entry/pfam/PF00002\n",
      "    https://www.ebi.ac.uk/interpro/entry/pfam/PF00003\n",
      "    https://www.ebi.ac.uk/interpro/entry/pfam/PF00004\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.org/dc/terms/description\n",
      "    http://purl.org/dc/terms/license\n",
      "    http://purl.org/dc/terms/title\n",
      "    http://www.geneontology.org/formats/oboInOwl#auto-generated-by\n",
      "    https://w3id.org/biopragmatics/resources/pfam/pfam.ofn\n",
      "\n",
      "File: po.owl\n",
      "  Has imports: Yes\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/po.owl\n",
      "  Own terms: 1996\n",
      "  External terms: 67\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 50\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000062\n",
      "    http://purl.obolibrary.org/obo/BFO_0000063\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/PO_0000001\n",
      "    http://purl.obolibrary.org/obo/PO_0000002\n",
      "    http://purl.obolibrary.org/obo/PO_0000003\n",
      "    http://purl.obolibrary.org/obo/PO_0000004\n",
      "    http://purl.obolibrary.org/obo/PO_0000005\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000062\n",
      "    http://purl.obolibrary.org/obo/BFO_0000063\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "\n",
      "File: rhea.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: https://w3id.org/biopragmatics/resources/rhea/rhea.ofn\n",
      "  Own terms: 0\n",
      "  External terms: 324997\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 68535\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/RO#reactionEnabledByMolecularFunction\n",
      "    http://purl.obolibrary.org/obo/RO_0000056\n",
      "    http://purl.obolibrary.org/obo/RO_0000057\n",
      "    http://purl.obolibrary.org/obo/RO_0002233\n",
      "    http://purl.obolibrary.org/obo/RO_0002234\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/CHEBI_100\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10015\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10023\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10036\n",
      "    http://purl.obolibrary.org/obo/CHEBI_10106\n",
      "\n",
      "File: ror.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: https://w3id.org/biopragmatics/resources/ror/ror.ofn\n",
      "  Own terms: 0\n",
      "  External terms: 122648\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 122639\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000062\n",
      "    http://purl.obolibrary.org/obo/BFO_0000063\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000856\n",
      "  First 5 own terms:\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BFO_0000050\n",
      "    http://purl.obolibrary.org/obo/BFO_0000051\n",
      "    http://purl.obolibrary.org/obo/BFO_0000062\n",
      "    http://purl.obolibrary.org/obo/BFO_0000063\n",
      "    http://purl.obolibrary.org/obo/ENVO_00000856\n",
      "\n",
      "File: so.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/so.owl\n",
      "  Own terms: 2639\n",
      "  External terms: 190\n",
      "  Classification: Potential Base (missing -base hint in .owl file name). No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 181\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/BS_00003\n",
      "    http://purl.obolibrary.org/obo/BS_0000338\n",
      "    http://purl.obolibrary.org/obo/BS_0000339\n",
      "    http://purl.obolibrary.org/obo/BS_0000340\n",
      "    http://purl.obolibrary.org/obo/BS_0000341\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/SO_0000000\n",
      "    http://purl.obolibrary.org/obo/SO_0000001\n",
      "    http://purl.obolibrary.org/obo/SO_0000002\n",
      "    http://purl.obolibrary.org/obo/SO_0000003\n",
      "    http://purl.obolibrary.org/obo/SO_0000004\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/BS_00003\n",
      "    http://purl.obolibrary.org/obo/BS_0000338\n",
      "    http://purl.obolibrary.org/obo/BS_0000339\n",
      "    http://purl.obolibrary.org/obo/BS_0000340\n",
      "    http://purl.obolibrary.org/obo/BS_0000341\n",
      "\n",
      "File: taxrank.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/taxrank.owl\n",
      "  Own terms: 74\n",
      "  External terms: 15\n",
      "  Classification: Non-Base. No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 10\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000112\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000700\n",
      "    http://purl.obolibrary.org/obo/taxrank.owl\n",
      "    http://www.geneontology.org/formats/oboInOwl#created_by\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000000\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000001\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000002\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000003\n",
      "    http://purl.obolibrary.org/obo/TAXRANK_0000004\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/IAO_0000112\n",
      "    http://purl.obolibrary.org/obo/IAO_0000115\n",
      "    http://purl.obolibrary.org/obo/IAO_0000700\n",
      "    http://purl.obolibrary.org/obo/taxrank.owl\n",
      "    http://purl.org/dc/terms/contributor\n",
      "\n",
      "File: uo.owl\n",
      "  Has imports: No\n",
      "  Ontology IRI: http://purl.obolibrary.org/obo/uo.owl\n",
      "  Own terms: 572\n",
      "  External terms: 5\n",
      "  Classification: Potential Base (missing -base hint in .owl file name). No -base version available in OBO Foundry\n",
      "  External Terms Subject of Triples? Yes\n",
      "  Number of external terms that are subjects of triples: 1\n",
      "  First 5 external terms that are subject of triples:\n",
      "    http://purl.obolibrary.org/obo/uo.owl\n",
      "  First 5 own terms:\n",
      "    http://purl.obolibrary.org/obo/UO_0000000\n",
      "    http://purl.obolibrary.org/obo/UO_0000001\n",
      "    http://purl.obolibrary.org/obo/UO_0000002\n",
      "    http://purl.obolibrary.org/obo/UO_0000003\n",
      "    http://purl.obolibrary.org/obo/UO_0000004\n",
      "  First 5 external terms:\n",
      "    http://purl.obolibrary.org/obo/has_prefix\n",
      "    http://purl.obolibrary.org/obo/is_unit_of\n",
      "    http://purl.obolibrary.org/obo/uo.owl\n",
      "    http://www.geneontology.org/formats/oboInOwl#hasExactSynonym\n",
      "    http://www.geneontology.org/formats/oboInOwl#hasRelatedSynonym\n",
      "\n",
      "Analysis complete!\n",
      "CPU times: user 1min, sys: 5.74 s, total: 1min 6s\n",
      "Wall time: 2min 4s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the analysis\n",
    "import analyze_non_core_ontologies\n",
    "\n",
    "# Run the analysis\n",
    "analyze_non_core_ontologies.analyze_non_core_ontologies(repo_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87eb0912-a5e9-4e2f-9362-37f1cc18b553",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Recreate pseudo base versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "704c982f-8ab8-47ca-9c95-bbdb1953c05a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /scratch/jplfaria/install_stuff/robot/robot\n",
      "Processing po.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/PO_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/po.owl --base-iri http://purl.obolibrary.org/obo/PO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl\n",
      "Created base version for po.owl: po-base.owl\n",
      "Processing iao.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/IAO_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/iao.owl --base-iri http://purl.obolibrary.org/obo/IAO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl\n",
      "Created base version for iao.owl: iao-base.owl\n",
      "Processing omo.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/OMO_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/omo.owl --base-iri http://purl.obolibrary.org/obo/OMO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl\n",
      "Created base version for omo.owl: omo-base.owl\n",
      "Processing bfo.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/BFO_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/bfo.owl --base-iri http://purl.obolibrary.org/obo/BFO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl\n",
      "Created base version for bfo.owl: bfo-base.owl\n",
      "Processing foodon.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/FOODON_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/foodon.owl --base-iri http://purl.obolibrary.org/obo/FOODON_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl\n",
      "Created base version for foodon.owl: foodon-base.owl\n",
      "Processing uo.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/UO_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/uo.owl --base-iri http://purl.obolibrary.org/obo/UO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl\n",
      "Created base version for uo.owl: uo-base.owl\n",
      "Processing taxrank.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/TAXRANK_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/taxrank.owl --base-iri http://purl.obolibrary.org/obo/TAXRANK_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl\n",
      "Created base version for taxrank.owl: taxrank-base.owl\n",
      "Processing so.owl...\n",
      "Using base IRI: http://purl.obolibrary.org/obo/SO_\n",
      "Executing command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/non-base-ontologies/so.owl --base-iri http://purl.obolibrary.org/obo/SO_ --axioms external --preserve-structure false --trim false remove --select imports --trim false --output /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl\n",
      "Created base version for so.owl: so-base.owl\n",
      "Copying PyOBO ontology to main directory: gtdb.owl\n",
      "Copying PyOBO ontology to main directory: pfam.owl\n",
      "Copying PyOBO ontology to main directory: eccode.owl\n",
      "Copying PyOBO ontology to main directory: credit.owl\n",
      "Copying PyOBO ontology to main directory: rhea.owl\n",
      "Copying PyOBO ontology to main directory: interpro.owl\n",
      "Copying PyOBO ontology to main directory: ror.owl\n",
      "\n",
      "Processing complete!\n",
      "CPU times: user 34.6 ms, sys: 834 ms, total: 868 ms\n",
      "Wall time: 44.9 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "# Import and run the pseudo base creation\n",
    "from create_pseudo_base_ontology import create_pseudo_base_ontologies\n",
    "# Run the creation\n",
    "create_pseudo_base_ontologies(repo_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29336d90-3d6e-4289-87ae-9afcdc41a3dd",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Analyze the prefixes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "97848151-5f93-4219-bca9-787261a5eef3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing ontologies in: /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_core\n",
      "\n",
      "Analyzing ncbitaxon.owl...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import the prefix analyzer\n",
    "from analyze_prefixes import analyze_all_ontologies, generate_prefix_mapping\n",
    "\n",
    "# Define input directory path\n",
    "input_dir = os.path.join(repo_path, 'ontology_data_owl_core')\n",
    "\n",
    "# Run the analysis\n",
    "print(f\"Analyzing ontologies in: {input_dir}\")\n",
    "results = analyze_all_ontologies(input_dir)\n",
    "\n",
    "# Generate and save prefix mapping\n",
    "mapping_content = generate_prefix_mapping(results)\n",
    "mapping_file = os.path.join(repo_path, 'prefix_mapping.txt')\n",
    "\n",
    "with open(mapping_file, 'w') as f:\n",
    "    f.write(mapping_content)\n",
    "\n",
    "print(f\"\\nPrefix mapping file generated at: {mapping_file}\")\n",
    "\n",
    "# Print summary of analysis\n",
    "print(\"\\nSummary of analysis:\")\n",
    "for filename, data in results.items():\n",
    "    print(f\"\\n{filename}:\")\n",
    "    print(f\"  Declared prefixes: {len(data['prefixes'])}\")\n",
    "    additional_prefixes = set(data['prefix_to_iris'].keys()) - data['prefixes']\n",
    "    print(f\"  Potential additional prefixes needed: {len(additional_prefixes)}\")\n",
    "    if additional_prefixes:\n",
    "        print(\"  Additional prefixes:\")\n",
    "        for prefix in sorted(additional_prefixes):\n",
    "            iris = data['prefix_to_iris'][prefix]\n",
    "            if iris:\n",
    "                print(f\"    - {prefix}: {next(iter(iris))}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "970c3736-eb5b-4461-9fe8-25c0e2b97d70",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Merge Ontologies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4add5f0c-5009-4998-896f-2b95759caa2f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting ontology merge - Check merge_memory.log for memory usage details\n",
      "Using ROBOT at: /scratch/jplfaria/install_stuff/robot/robot\n",
      "Looking for ontology files in: /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl\n",
      "Found 30 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl.owl\n",
      "\n",
      "Merge completed - Peak memory usage: 1331GB\n",
      "CPU times: user 497 ms, sys: 369 ms, total: 866 ms\n",
      "Wall time: 1h 26min 3s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "from threading import Thread\n",
    "import time\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    filename='merge_memory.log',\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(message)s'\n",
    ")\n",
    "\n",
    "# Initialize peak memory tracking\n",
    "peak_memory_gb = 0\n",
    "last_logged_memory = 0\n",
    "\n",
    "def parse_memory_output(mem_line):\n",
    "    parts = mem_line.split()\n",
    "    used_mem = parts[2]\n",
    "    value = float(used_mem[:-2])\n",
    "    return value * 1024 if used_mem.endswith('Ti') else value\n",
    "\n",
    "def monitor_memory():\n",
    "    global peak_memory_gb, last_logged_memory\n",
    "    while True:\n",
    "        try:\n",
    "            mem_info = subprocess.check_output(['free', '-h']).decode('utf-8').split('\\n')\n",
    "            current_memory_gb = parse_memory_output(mem_info[1])\n",
    "            peak_memory_gb = max(peak_memory_gb, current_memory_gb)\n",
    "            \n",
    "            # Log if memory changed by more than 10GB\n",
    "            if abs(current_memory_gb - last_logged_memory) > 10:\n",
    "                logging.info(f\"Memory: {current_memory_gb:.0f}GB (Peak: {peak_memory_gb:.0f}GB)\")\n",
    "                last_logged_memory = current_memory_gb\n",
    "            \n",
    "            time.sleep(60)\n",
    "        except:\n",
    "            break\n",
    "\n",
    "# Start memory monitoring in background\n",
    "monitor_thread = Thread(target=monitor_memory, daemon=True)\n",
    "monitor_thread.start()\n",
    "\n",
    "# Run merge\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "from merge_ontologies import merge_ontologies\n",
    "\n",
    "print(\"Starting ontology merge - Check merge_memory.log for memory usage details\")\n",
    "logging.info(\"Starting ontology merge\")\n",
    "\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',\n",
    "    output_filename='ontology_data_owl.owl'\n",
    ")\n",
    "\n",
    "# Log final statistics\n",
    "logging.info(f\"Merge completed - Final peak memory: {peak_memory_gb:.0f}GB\")\n",
    "print(f\"\\nMerge completed - Peak memory usage: {peak_memory_gb:.0f}GB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "677d384f-9269-4665-b38e-171c8c21a929",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting ontology merge - Check merge_memory.log for memory usage details\n",
      "Found 30 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "\n",
      "Executing initial merge command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/intermediary_ontology/merged_intermediate.owl\n",
      "\n",
      "Executing annotate command:\n",
      "robot annotate --input /scratch/jplfaria/KBase_CDM_Ontologies/outputs/intermediary_ontology/merged_intermediate.owl --annotation-file /scratch/jplfaria/KBase_CDM_Ontologies/outputs/intermediary_ontology/prefix_mapping.txt --prefix EC: http://purl.obolibrary.org/obo/EC_ --prefix eccode: http://purl.obolibrary.org/obo/EC_ --prefix UP: http://purl.uniprot.org/uniprot/ --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/intermediary_ontology/prefix_fixed.owl\n",
      "\n",
      "Executing remove command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/outputs/intermediary_ontology/prefix_fixed.owl --term oio:id --select annotation-properties --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_prefix_EC_UP.owl\n",
      "\n",
      "Successfully merged ontologies into /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_prefix_EC_UP.owl\n",
      "\n",
      "Merge completed - Peak memory usage: 1331GB\n",
      "CPU times: user 983 ms, sys: 630 ms, total: 1.61 s\n",
      "Wall time: 3h 44min 30s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "from threading import Thread\n",
    "import time\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    filename='merge_memory.log',\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(message)s'\n",
    ")\n",
    "\n",
    "# Initialize peak memory tracking\n",
    "peak_memory_gb = 0\n",
    "last_logged_memory = 0\n",
    "\n",
    "def parse_memory_output(mem_line):\n",
    "    parts = mem_line.split()\n",
    "    used_mem = parts[2]\n",
    "    value = float(used_mem[:-2])\n",
    "    return value * 1024 if used_mem.endswith('Ti') else value\n",
    "\n",
    "def monitor_memory():\n",
    "    global peak_memory_gb, last_logged_memory\n",
    "    while True:\n",
    "        try:\n",
    "            mem_info = subprocess.check_output(['free', '-h']).decode('utf-8').split('\\n')\n",
    "            current_memory_gb = parse_memory_output(mem_info[1])\n",
    "            peak_memory_gb = max(peak_memory_gb, current_memory_gb)\n",
    "            \n",
    "            # Log if memory changed by more than 10GB\n",
    "            if abs(current_memory_gb - last_logged_memory) > 10:\n",
    "                logging.info(f\"Memory: {current_memory_gb:.0f}GB (Peak: {peak_memory_gb:.0f}GB)\")\n",
    "                last_logged_memory = current_memory_gb\n",
    "            \n",
    "            time.sleep(60)\n",
    "        except:\n",
    "            break\n",
    "\n",
    "# Start memory monitoring in background\n",
    "monitor_thread = Thread(target=monitor_memory, daemon=True)\n",
    "monitor_thread.start()\n",
    "\n",
    "# Run merge\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "from merge_ontologies_v2 import merge_ontologies\n",
    "\n",
    "print(\"Starting ontology merge - Check merge_memory.log for memory usage details\")\n",
    "logging.info(\"Starting ontology merge\")\n",
    "\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',\n",
    "    output_filename='ontology_data_owl_prefix_EC_UP.owl'\n",
    ")\n",
    "\n",
    "# Log final statistics\n",
    "logging.info(f\"Merge completed - Final peak memory: {peak_memory_gb:.0f}GB\")\n",
    "print(f\"\\nMerge completed - Peak memory usage: {peak_memory_gb:.0f}GB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "57db83e7-6c74-4cc7-8565-e279c0c9bbf4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting ontology merge - Check merge_memory.log for memory usage details\n",
      "Found 30 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "\n",
      "Executing initial merge command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/intermediary_ontology/merged_intermediate.owl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "from threading import Thread\n",
    "import time\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    filename='merge_memory.log',\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(message)s'\n",
    ")\n",
    "\n",
    "# Initialize peak memory tracking\n",
    "peak_memory_gb = 0\n",
    "last_logged_memory = 0\n",
    "\n",
    "def parse_memory_output(mem_line):\n",
    "    parts = mem_line.split()\n",
    "    used_mem = parts[2]\n",
    "    value = float(used_mem[:-2])\n",
    "    return value * 1024 if used_mem.endswith('Ti') else value\n",
    "\n",
    "def monitor_memory():\n",
    "    global peak_memory_gb, last_logged_memory\n",
    "    while True:\n",
    "        try:\n",
    "            mem_info = subprocess.check_output(['free', '-h']).decode('utf-8').split('\\n')\n",
    "            current_memory_gb = parse_memory_output(mem_info[1])\n",
    "            peak_memory_gb = max(peak_memory_gb, current_memory_gb)\n",
    "            \n",
    "            # Log if memory changed by more than 10GB\n",
    "            if abs(current_memory_gb - last_logged_memory) > 10:\n",
    "                logging.info(f\"Memory: {current_memory_gb:.0f}GB (Peak: {peak_memory_gb:.0f}GB)\")\n",
    "                last_logged_memory = current_memory_gb\n",
    "            \n",
    "            time.sleep(60)\n",
    "        except:\n",
    "            break\n",
    "\n",
    "# Start memory monitoring in background\n",
    "monitor_thread = Thread(target=monitor_memory, daemon=True)\n",
    "monitor_thread.start()\n",
    "\n",
    "# Run merge\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "from merge_ontologies_v3 import merge_ontologies\n",
    "\n",
    "print(\"Starting ontology merge - Check merge_memory.log for memory usage details\")\n",
    "logging.info(\"Starting ontology merge\")\n",
    "\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',\n",
    "    output_filename='ontology_data_owl_set_prefixes.owl'\n",
    ")\n",
    "\n",
    "# Log final statistics\n",
    "logging.info(f\"Merge completed - Final peak memory: {peak_memory_gb:.0f}GB\")\n",
    "print(f\"\\nMerge completed - Peak memory usage: {peak_memory_gb:.0f}GB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4e54221e-f015-4ad5-ab72-c6a5abd10153",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Executing remove command:\n",
      "robot remove --input /scratch/jplfaria/KBase_CDM_Ontologies/outputs/intermediary_ontology/prefix_fixed.owl --term oio:id --select annotation-properties --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_set_prefixes.owl\n",
      "\n",
      "Successfully created final ontology at /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_set_prefixes.owl\n",
      "CPU times: user 222 ms, sys: 191 ms, total: 414 ms\n",
      "Wall time: 1h 7min 21s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "output_dir = os.path.join(repo_path, 'outputs')\n",
    "intermediate_dir = os.path.join(output_dir, 'intermediary_ontology')\n",
    "\n",
    "# Input file (your existing prefix_fixed.owl)\n",
    "intermediate_file2 = os.path.join(intermediate_dir, 'prefix_fixed.owl')\n",
    "\n",
    "# Output file\n",
    "output_file = os.path.join(output_dir, 'ontology_data_owl_set_prefixes.owl')\n",
    "\n",
    "# Step 3: Remove annotations using filter\n",
    "filter_command = [\n",
    "    'robot', 'remove',\n",
    "    '--input', intermediate_file2,\n",
    "    '--term', 'oio:id',\n",
    "    '--select', 'annotation-properties',\n",
    "    '--output', output_file\n",
    "]\n",
    "\n",
    "print(\"\\nExecuting remove command:\")\n",
    "print(f\"{' '.join(filter_command)}\")\n",
    "\n",
    "try:\n",
    "    subprocess.run(filter_command, check=True, capture_output=True, text=True)\n",
    "    print(f\"\\nSuccessfully created final ontology at {output_file}\")\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"\\nError executing ROBOT command:\")\n",
    "    print(f\"Return code: {e.returncode}\")\n",
    "    if e.stdout:\n",
    "        print(\"STDOUT:\", e.stdout)\n",
    "    if e.stderr:\n",
    "        print(\"STDERR:\", e.stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31b9153d-08c7-4ece-a45d-7bc22e342e24",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the merge\n",
    "from merge_ontologies_v4 import merge_ontologies\n",
    "\n",
    "# Run the merge\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl_test',  # Your input directory\n",
    "    output_filename='test_merged_ontologies.owl'  # Your output filename\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "12d18b08-b02e-43be-aaf5-2c0f5ff07055",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Analyzing individual ontologies...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl...\n",
      "Processing /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl...\n",
      "Individual ontology prefixes saved to /scratch/jplfaria/KBase_CDM_Ontologies/outputs/prefix_analysis/individual_prefixes.csv\n",
      "Combined prefixes saved to /scratch/jplfaria/KBase_CDM_Ontologies/outputs/prefix_analysis/combined_prefixes.csv\n",
      "\n",
      "Analyzing merged ontology...\n",
      "Merged ontology prefixes saved to /scratch/jplfaria/KBase_CDM_Ontologies/outputs/prefix_analysis/merged_prefixes.csv\n",
      "\n",
      "Prefix Analysis:\n",
      "\n",
      "Total prefixes in individual ontologies: 109\n",
      "Total prefixes in merged ontology: 59\n",
      "\n",
      "Prefixes lost after merging (54):\n",
      "  - BFO: http://purl.obolibrary.org/obo/BFO_\n",
      "  - CHEBI: http://purl.obolibrary.org/obo/CHEBI_\n",
      "  - ENVO: http://purl.obolibrary.org/obo/ENVO_\n",
      "  - GO: http://purl.obolibrary.org/obo/GO_\n",
      "  - IAO: http://purl.obolibrary.org/obo/IAO_\n",
      "  - NCBITaxon: http://purl.obolibrary.org/obo/NCBITaxon_\n",
      "  - OBI: http://purl.obolibrary.org/obo/OBI_\n",
      "  - OMO: http://purl.obolibrary.org/obo/OMO_\n",
      "  - RO: http://purl.obolibrary.org/obo/RO_\n",
      "  - TAXRANK: http://purl.obolibrary.org/obo/TAXRANK_\n",
      "  - asfis: https://www.fao.org/fishery/en/aqspecies/\n",
      "  - biolink: https://w3id.org/biolink/vocab/\n",
      "  - chebi1: http://purl.obolibrary.org/obo/chebi#2\n",
      "  - chebi2: http://purl.obolibrary.org/obo/chebi#\n",
      "  - cnrs: https://bioregistry.io/cnrs:\n",
      "  - credit: https://credit.niso.org/contributor-roles/\n",
      "  - debio: https://biopragmatics.github.io/debio/\n",
      "  - eccode: https://www.ebi.ac.uk/intenz/query?cmd=SearchEC&ec=\n",
      "  - ecocyc: https://ecocyc.org/gene?id=\n",
      "  - efo: http://www.ebi.ac.uk/efo/\n",
      "  - eolife: http://eol.org/pages/\n",
      "  - funderregistry: http://data.crossref.org/fundingdata/funder/\n",
      "  - geonames: https://www.geonames.org/\n",
      "  - grid: https://www.grid.ac/institutes/\n",
      "  - grin: http://www.ars-grin.gov/cgi-bin/npgs/html/taxon.pl?\n",
      "  - gtdb: https://gtdb.ecogenomic.org/tree?r=\n",
      "  - hesa: https://bioregistry.io/hesa:\n",
      "  - indexfungorum: http://www.indexfungorum.org/names/NamesRecord.asp?RecordID=\n",
      "  - interpro: https://www.ebi.ac.uk/interpro/entry/InterPro/\n",
      "  - isni: http://www.isni.org/isni/\n",
      "  - itis: http://www.itis.gov/servlet/SingleRpt/SingleRpt?search_topic=TSN&search_value=\n",
      "  - kegg.reaction: https://www.kegg.jp/entry/\n",
      "  - langual: http://www.langual.org/langual_thesaurus.asp?termid=\n",
      "  - macie: https://www.ebi.ac.uk/thornton-srv/m-csa/search/?s=\n",
      "  - metacyc.compound: https://metacyc.org/compound?orgid=META&id=\n",
      "  - obo1: http://www.geneontology.org/formats/oboInOwl#http://purl.obolibrary.org/obo/\n",
      "  - orcid: https://orcid.org/\n",
      "  - pfam: https://www.ebi.ac.uk/interpro/entry/pfam/\n",
      "  - pfam.clan: https://www.ebi.ac.uk/interpro/set/pfam/\n",
      "  - po: http://purl.obolibrary.org/obo/po#\n",
      "  - pr: http://purl.obolibrary.org/obo/pr#\n",
      "  - reactome: https://reactome.org/content/detail/\n",
      "  - rhea1: http://purl.obolibrary.org/obo/rhea#\n",
      "  - ror: https://ror.org/\n",
      "  - siren: https://bioregistry.io/reference/siren:\n",
      "  - so: http://purl.obolibrary.org/obo/so#\n",
      "  - subsets: http://purl.obolibrary.org/obo/ro/subsets#\n",
      "  - uberon: http://purl.obolibrary.org/obo/uberon#\n",
      "  - ubprop: http://purl.obolibrary.org/obo/ubprop#\n",
      "  - ucas: https://bioregistry.io/ucas:\n",
      "  - ukprn: https://bioregistry.io/ukprn:\n",
      "  - uniprot: http://purl.uniprot.org/uniprot/\n",
      "  - wikidata: http://www.wikidata.org/entity/\n",
      "  - wikipedia: https://en.wikipedia.org/wiki/\n",
      "\n",
      "New prefixes in merged ontology (4):\n",
      "  - chebi5: http://purl.obolibrary.org/obo/chebi#3\n",
      "  - chebi6: http://purl.obolibrary.org/obo/chebi#1\n",
      "  - obo2: http://www.geneontology.org/formats/oboInOwl#http://purl.obolibrary.org/obo/\n",
      "  - vocab: https://w3id.org/biolink/vocab/\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from rdflib import Graph\n",
    "import csv\n",
    "from typing import Dict, Tuple\n",
    "\n",
    "def extract_prefixes_from_ontology(file_path: str) -> dict:\n",
    "    \"\"\"Extract prefixes from a single ontology file.\"\"\"\n",
    "    g = Graph()\n",
    "    g.parse(file_path, format=\"xml\" if file_path.endswith(\".owl\") else \"turtle\")\n",
    "    return {prefix: str(namespace) for prefix, namespace in g.namespaces()}\n",
    "\n",
    "def extract_prefixes_from_directory(directory: str) -> Tuple[Dict[str, Dict[str, str]], Dict[str, str]]:\n",
    "    \"\"\"\n",
    "    Extract prefixes from all ontology files in a directory and track per-file prefixes.\n",
    "    \n",
    "    Returns:\n",
    "        Tuple containing:\n",
    "        - Dictionary mapping filenames to their prefix dictionaries\n",
    "        - Combined dictionary of all prefixes\n",
    "    \"\"\"\n",
    "    per_file_prefixes = {}\n",
    "    all_prefixes = {}\n",
    "    \n",
    "    for filename in os.listdir(directory):\n",
    "        if filename.endswith((\".owl\", \".ttl\", \".ofn\", \".rdf\")):\n",
    "            file_path = os.path.join(directory, filename)\n",
    "            print(f\"Processing {file_path}...\")\n",
    "            try:\n",
    "                file_prefixes = extract_prefixes_from_ontology(file_path)\n",
    "                per_file_prefixes[filename] = file_prefixes\n",
    "                all_prefixes.update(file_prefixes)\n",
    "            except Exception as e:\n",
    "                print(f\"Error processing {file_path}: {e}\")\n",
    "    \n",
    "    return per_file_prefixes, all_prefixes\n",
    "\n",
    "def save_prefixes_to_csv(prefixes: dict, output_file: str, include_source: bool = False):\n",
    "    \"\"\"Save prefixes to a CSV file.\"\"\"\n",
    "    with open(output_file, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile)\n",
    "        if include_source:\n",
    "            writer.writerow([\"Source\", \"Prefix\", \"Namespace\"])\n",
    "            for source, prefix_dict in prefixes.items():\n",
    "                for prefix, namespace in sorted(prefix_dict.items()):\n",
    "                    writer.writerow([source, prefix, namespace])\n",
    "        else:\n",
    "            writer.writerow([\"Prefix\", \"Namespace\"])\n",
    "            for prefix, namespace in sorted(prefixes.items()):\n",
    "                writer.writerow([prefix, namespace])\n",
    "\n",
    "def analyze_prefixes(input_dir: str, merged_file: str, output_dir: str):\n",
    "    \"\"\"\n",
    "    Analyze prefixes in both individual ontologies and merged ontology.\n",
    "    \n",
    "    Args:\n",
    "        input_dir: Directory containing individual ontologies\n",
    "        merged_file: Path to the merged ontology file\n",
    "        output_dir: Directory for output files\n",
    "    \"\"\"\n",
    "    # Create output directory if it doesn't exist\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    # Analyze individual ontologies\n",
    "    print(\"\\nAnalyzing individual ontologies...\")\n",
    "    per_file_prefixes, all_prefixes = extract_prefixes_from_directory(input_dir)\n",
    "    \n",
    "    # Save individual ontology prefixes\n",
    "    individual_output = os.path.join(output_dir, \"individual_prefixes.csv\")\n",
    "    save_prefixes_to_csv(per_file_prefixes, individual_output, include_source=True)\n",
    "    print(f\"Individual ontology prefixes saved to {individual_output}\")\n",
    "    \n",
    "    # Save combined prefixes from all individual ontologies\n",
    "    combined_output = os.path.join(output_dir, \"combined_prefixes.csv\")\n",
    "    save_prefixes_to_csv(all_prefixes, combined_output)\n",
    "    print(f\"Combined prefixes saved to {combined_output}\")\n",
    "    \n",
    "    # Analyze merged ontology\n",
    "    if os.path.exists(merged_file):\n",
    "        print(\"\\nAnalyzing merged ontology...\")\n",
    "        try:\n",
    "            merged_prefixes = extract_prefixes_from_ontology(merged_file)\n",
    "            merged_output = os.path.join(output_dir, \"merged_prefixes.csv\")\n",
    "            save_prefixes_to_csv(merged_prefixes, merged_output)\n",
    "            print(f\"Merged ontology prefixes saved to {merged_output}\")\n",
    "            \n",
    "            # Compare prefixes\n",
    "            print(\"\\nPrefix Analysis:\")\n",
    "            original_prefixes = set(all_prefixes.keys())\n",
    "            merged_prefix_set = set(merged_prefixes.keys())\n",
    "            \n",
    "            print(f\"\\nTotal prefixes in individual ontologies: {len(original_prefixes)}\")\n",
    "            print(f\"Total prefixes in merged ontology: {len(merged_prefix_set)}\")\n",
    "            \n",
    "            lost_prefixes = original_prefixes - merged_prefix_set\n",
    "            new_prefixes = merged_prefix_set - original_prefixes\n",
    "            \n",
    "            if lost_prefixes:\n",
    "                print(f\"\\nPrefixes lost after merging ({len(lost_prefixes)}):\")\n",
    "                for prefix in sorted(lost_prefixes):\n",
    "                    print(f\"  - {prefix}: {all_prefixes[prefix]}\")\n",
    "            \n",
    "            if new_prefixes:\n",
    "                print(f\"\\nNew prefixes in merged ontology ({len(new_prefixes)}):\")\n",
    "                for prefix in sorted(new_prefixes):\n",
    "                    print(f\"  - {prefix}: {merged_prefixes[prefix]}\")\n",
    "                    \n",
    "        except Exception as e:\n",
    "            print(f\"Error analyzing merged ontology: {e}\")\n",
    "    else:\n",
    "        print(f\"\\nMerged ontology file not found: {merged_file}\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # Configuration\n",
    "    input_dir = \"/scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl\"\n",
    "    merged_file = \"/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl.owl\"\n",
    "    output_dir = \"/scratch/jplfaria/KBase_CDM_Ontologies/outputs/prefix_analysis\"\n",
    "    \n",
    "    analyze_prefixes(input_dir, merged_file, output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "18743f68-0fb9-4a98-9fb0-e10bfeab4a34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /cdm_shared_workspace/user_shared_workspace/Jose/install_stuff/robot/robot\n",
      "Looking for ontology files in: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core\n",
      "Found 4 ontology files:\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/ncbitaxon.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/envo.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/chebi.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/go.owl\n",
      "Created list of merged ontologies at: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/ontology_data_owl_core.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/ncbitaxon.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/envo.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/chebi.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core/go.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/ontology_data_owl_core.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/ontology_data_owl_core.owl\n",
      "CPU times: user 1.28 s, sys: 547 ms, total: 1.83 s\n",
      "Wall time: 35min 27s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the merge\n",
    "from merge_ontologies import merge_ontologies\n",
    "\n",
    "# Run the merge with custom input directory and output filename\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl_core',  # Custom input directory\n",
    "    output_filename='ontology_data_owl_core.owl'    # Custom output filename\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d450ca66-2119-44e3-999c-d0cfe2cc545a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /scratch/jplfaria/install_stuff/robot/robot\n",
      "Looking for ontology files in: /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl\n",
      "Found 31 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/seed.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/bfo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/caro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/cl-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/fao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/foodon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/iao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/obi-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/omo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pato-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pco-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/po-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/so-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/taxrank-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uberon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/uo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/pfam.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/credit.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ror.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/interpro.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/seed.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31.owl\n",
      "CPU times: user 706 ms, sys: 216 ms, total: 922 ms\n",
      "Wall time: 1h 12min 20s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import the merge function\n",
    "from merge_ontologies_custom_order import merge_ontologies\n",
    "\n",
    "# Define the desired order for merging (filenames must be present in the input directory)\n",
    "desired_order = [\n",
    "    'ncbitaxon.owl',\n",
    "    'go.owl',\n",
    "    'envo.owl',\n",
    "    'chebi.owl',\n",
    "    'bfo-base.owl',\n",
    "    'caro-base.owl',\n",
    "    'cl-base.owl',\n",
    "    'fao-base.owl',\n",
    "    'foodon-base.owl',\n",
    "    'iao-base.owl',\n",
    "    'obi-base.owl',\n",
    "    'omo-base.owl',\n",
    "    'pato-base.owl',\n",
    "    'pco-base.owl',\n",
    "    'po-base.owl',\n",
    "    'ro-base.owl',\n",
    "    'so-base.owl',\n",
    "    'taxrank-base.owl',\n",
    "    'uberon-base.owl',\n",
    "    'uo-base.owl',\n",
    "    'gtdb.owl',\n",
    "    'eccode.owl',\n",
    "    'pfam.owl',\n",
    "    'rhea.owl',\n",
    "    'credit.owl',\n",
    "    'ror.owl',\n",
    "    'interpro.owl',\n",
    "    'seed.owl',\n",
    "    'modelseed.owl',\n",
    "    'kegg.owl',\n",
    "    'metacyc.owl'\n",
    "]\n",
    "\n",
    "# Run the merge with custom input directory, output filename, and the specified file order\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',  # Custom input directory\n",
    "    output_filename='ontology_data_owl_31.owl',  # Custom output filename\n",
    "    ontology_order=desired_order\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8f16b5ea-845c-4228-b498-bf994244e12e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /scratch/jplfaria/install_stuff/robot/robot\n",
      "Looking for ontology files in: /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl\n",
      "Found 7 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/7_biochemistry.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/kegg.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/metacyc.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/7_biochemistry.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /scratch/jplfaria/KBase_CDM_Ontologies/outputs/7_biochemistry.owl\n",
      "CPU times: user 86 ms, sys: 48 ms, total: 134 ms\n",
      "Wall time: 11min 55s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import the merge function\n",
    "from merge_ontologies_custom_order import merge_ontologies\n",
    "\n",
    "# Define the desired order for merging (filenames must be present in the input directory)\n",
    "desired_order = [\n",
    "    'go.owl',\n",
    "    'chebi.owl',\n",
    "    'eccode.owl',\n",
    "    'rhea.owl',\n",
    "    'modelseed.owl',\n",
    "    'kegg.owl',\n",
    "    'metacyc.owl'\n",
    "]\n",
    "\n",
    "# Run the merge with custom input directory, output filename, and the specified file order\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',  # Custom input directory\n",
    "    output_filename='7_biochemistry.owl',  # Custom output filename\n",
    "    ontology_order=desired_order\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d6bd7c4f-a210-4f2a-beb1-5384e79dde6f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /scratch/jplfaria/install_stuff/robot/robot\n",
      "Looking for ontology files in: /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl\n",
      "Found 8 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/core_plus_ec_rhea_gtdb_modelseed.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/envo.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/chebi.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/gtdb.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/rhea.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/eccode.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/modelseed.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/core_plus_ec_rhea_gtdb_modelseed.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /scratch/jplfaria/KBase_CDM_Ontologies/outputs/core_plus_ec_rhea_gtdb_modelseed.owl\n",
      "CPU times: user 320 ms, sys: 255 ms, total: 574 ms\n",
      "Wall time: 58min 57s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import the merge function\n",
    "from merge_ontologies_custom_order import merge_ontologies\n",
    "\n",
    "# Define the desired order for merging (filenames must be present in the input directory)\n",
    "desired_order = [\n",
    "    'ncbitaxon.owl',\n",
    "    'go.owl',\n",
    "    'envo.owl',\n",
    "    'chebi.owl',\n",
    "    'gtdb.owl',\n",
    "    'rhea.owl',\n",
    "    'eccode.owl',\n",
    "    'modelseed.owl'\n",
    "]\n",
    "\n",
    "# Run the merge with custom input directory, output filename, and the specified file order\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',  # Custom input directory\n",
    "    output_filename='core_plus_ec_rhea_gtdb_modelseed.owl',  # Custom output filename\n",
    "    ontology_order=desired_order\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1bbe754e-05f5-4e4c-996a-fd3675af19a6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /scratch/jplfaria/install_stuff/robot/robot\n",
      "Looking for ontology files in: /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl\n",
      "Found 2 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ncbi_go.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/ncbitaxon.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl/go.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ncbi_go.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ncbi_go.owl\n",
      "CPU times: user 309 ms, sys: 168 ms, total: 478 ms\n",
      "Wall time: 50min 36s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import the merge function\n",
    "from merge_ontologies_custom_order import merge_ontologies\n",
    "\n",
    "# Define the desired order for merging (filenames must be present in the input directory)\n",
    "desired_order = [\n",
    "    'ncbitaxon.owl',\n",
    "    'go.owl'\n",
    "]\n",
    "\n",
    "# Run the merge with custom input directory, output filename, and the specified file order\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',  # Custom input directory\n",
    "    output_filename='ncbi_go.owl',  # Custom output filename\n",
    "    ontology_order=desired_order\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "893dc9a8-8a1d-4648-b9b4-94b0e91027f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "from threading import Thread\n",
    "import time\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    filename='merge_memory.log',\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(message)s'\n",
    ")\n",
    "\n",
    "# Initialize peak memory tracking\n",
    "peak_memory_gb = 0\n",
    "last_logged_memory = 0\n",
    "\n",
    "def parse_memory_output(mem_line):\n",
    "    parts = mem_line.split()\n",
    "    used_mem = parts[2]\n",
    "    value = float(used_mem[:-2])\n",
    "    return value * 1024 if used_mem.endswith('Ti') else value\n",
    "\n",
    "def monitor_memory():\n",
    "    global peak_memory_gb, last_logged_memory\n",
    "    while True:\n",
    "        try:\n",
    "            mem_info = subprocess.check_output(['free', '-h']).decode('utf-8').split('\\n')\n",
    "            current_memory_gb = parse_memory_output(mem_info[1])\n",
    "            peak_memory_gb = max(peak_memory_gb, current_memory_gb)\n",
    "            \n",
    "            # Log if memory changed by more than 10GB\n",
    "            if abs(current_memory_gb - last_logged_memory) > 10:\n",
    "                logging.info(f\"Memory: {current_memory_gb:.0f}GB (Peak: {peak_memory_gb:.0f}GB)\")\n",
    "                last_logged_memory = current_memory_gb\n",
    "            \n",
    "            time.sleep(60)\n",
    "        except:\n",
    "            break\n",
    "\n",
    "# Start memory monitoring in background\n",
    "monitor_thread = Thread(target=monitor_memory, daemon=True)\n",
    "monitor_thread.start()\n",
    "\n",
    "# Run merge\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "from merge_ontologies_v2 import merge_ontologies\n",
    "\n",
    "print(\"Starting ontology merge - Check merge_memory.log for memory usage details\")\n",
    "logging.info(\"Starting ontology merge\")\n",
    "\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl',\n",
    "    output_filename='ontology_data_owl_no_eccode.owl'\n",
    ")\n",
    "\n",
    "# Log final statistics\n",
    "logging.info(f\"Merge completed - Final peak memory: {peak_memory_gb:.0f}GB\")\n",
    "print(f\"\\nMerge completed - Peak memory usage: {peak_memory_gb:.0f}GB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2eba96ca-b998-43c7-b5aa-e1e134b959e2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /cdm_shared_workspace/user_shared_workspace/Jose/install_stuff/robot/robot\n",
      "Looking for ontology files in: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec\n",
      "Found 5 ontology files:\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/ncbitaxon.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/envo.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/chebi.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/eccode.owl\n",
      "  - /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/go.owl\n",
      "Created list of merged ontologies at: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/ontology_data_owl_core_ec.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/ncbitaxon.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/envo.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/chebi.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/eccode.owl --input /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/ontology_data_owl_core_ec/go.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/ontology_data_owl_core_ec.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/ontology_data_owl_core_ec.owl\n",
      "CPU times: user 6.49 s, sys: 0 ns, total: 6.49 s\n",
      "Wall time: 1h 10min 13s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the merge\n",
    "from merge_ontologies import merge_ontologies\n",
    "\n",
    "# Run the merge with custom input directory and output filename\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl_core_ec',  # Custom input directory\n",
    "    output_filename='ontology_data_owl_core_ec.owl'    # Custom output filename\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c31ea79-8de3-46ca-8193-5ac32d0c6164",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using ROBOT at: /scratch/jplfaria/install_stuff/robot/robot\n",
      "Looking for ontology files in: /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base\n",
      "Found 14 ontology files:\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/foodon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/cl-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/pco-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/iao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/caro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/uberon-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/uo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/omo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/bfo-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/po-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/ro-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/obi-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/fao-base.owl\n",
      "  - /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/pato-base.owl\n",
      "Created list of merged ontologies at: /scratch/jplfaria/KBase_CDM_Ontologies/ontologies_merged.txt\n",
      "Saving output to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_base.owl\n",
      "\n",
      "Executing command:\n",
      "robot merge --annotate-defined-by true --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/foodon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/cl-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/pco-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/iao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/caro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/uberon-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/uo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/omo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/bfo-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/po-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/ro-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/obi-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/fao-base.owl --input /scratch/jplfaria/KBase_CDM_Ontologies/ontology_data_owl_base/pato-base.owl remove --axioms disjoint --trim true --preserve-structure false remove --term owl:Nothing --trim true --preserve-structure false --output /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_base.owl\n",
      "\n",
      "ROBOT Output:\n",
      "Successfully merged ontologies into /scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_base.owl\n",
      "CPU times: user 11 ms, sys: 6.37 ms, total: 17.4 ms\n",
      "Wall time: 56.8 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the merge\n",
    "from merge_ontologies import merge_ontologies\n",
    "\n",
    "# Run the merge with custom input directory and output filename\n",
    "merge_ontologies(\n",
    "    repo_path,\n",
    "    input_dir_name='ontology_data_owl_base',  # Custom input directory\n",
    "    output_filename='ontology_data_owl_base.owl'    # Custom output filename\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9536cfaf-d481-4178-b12f-40f0ab1d24b8",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Create Semantic SQL DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9e34d064-8e8c-4b0b-839a-bf68fad15326",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current Python executable: /home/jplfaria/miniconda3/bin/python\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(f\"Current Python executable: {sys.executable}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "18854439-deba-4a5d-a494-c8899d2cabf7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/local/bin/python\n"
     ]
    }
   ],
   "source": [
    "!which python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "14759b18-0ce7-4424-b7c6-e77dc4480429",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error occurred: Could not find semsql installation location\n",
      "CPU times: user 24.9 ms, sys: 128 μs, total: 25.1 ms\n",
      "Wall time: 398 ms\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/scratch/jplfaria/KBase_CDM_Ontologies/scripts/create_semantic_sql_db.py\", line 24, in create_semantic_sql_db\n",
      "    raise Exception(\"Could not find semsql installation location\")\n",
      "Exception: Could not find semsql installation location\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the database creation\n",
    "from create_semantic_sql_db import create_semantic_sql_db\n",
    "\n",
    "# Run the database creation with custom input filename\n",
    "create_semantic_sql_db(\n",
    "    repo_path,\n",
    "    input_owl_filename='eccode.owl'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "018db057-28e2-4ad1-8fa6-12b7dc25abff",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input OWL file: ontology_data_owl.owl\n",
      "Output DB file: ontology_data_owl.db\n",
      "Working directory: /scratch/jplfaria/KBase_CDM_Ontologies/outputs\n",
      "\n",
      "Executing command:\n",
      "\n",
      "        export PATH=\"/scratch/jplfaria/bin:$PATH\"\n",
      "        source /home/jplfaria/scratch/install_stuff/setup_oak_env.sh\n",
      "        /home/jplfaria/scratch/install_stuff/semsql/bin/semsql make ontology_data_owl.db\n",
      "        \n",
      "\n",
      "Output:\n",
      "robot \\\n",
      "remove -i ontology_data_owl.owl --axioms \"equivalent disjoint annotation abox type\" \\\n",
      "filter --exclude-terms /home/jplfaria/scratch/install_stuff/semsql/semsql/builder//exclude-terms.txt \\\n",
      "-o ontology_data_owl-min.owl\n",
      "touch ontology_data_owl-properties.txt\n",
      "grep -v ^prefix, /home/jplfaria/scratch/install_stuff/semsql/semsql/builder/prefixes/prefixes.csv | grep -v ^obo, | perl -npe 's@,(.*)@: \"$1\"@' > ontology_data_owl-prefixes.yaml.tmp && mv ontology_data_owl-prefixes.yaml.tmp ontology_data_owl-prefixes.yaml\n",
      "relation-graph --disable-owl-nothing true \\\n",
      "--ontology-file ontology_data_owl-min.owl \\\n",
      " \\\n",
      "--output-file ontology_data_owl-relation-graph.tsv.tmp \\\n",
      "--equivalence-as-subclass true \\\n",
      "--mode TSV \\\n",
      "--prefixes ontology_data_owl-prefixes.yaml \\\n",
      "--output-individuals true \\\n",
      "--output-subclasses true \\\n",
      "--reflexive-subclasses true && \\\n",
      "mv ontology_data_owl-relation-graph.tsv.tmp ontology_data_owl-relation-graph.tsv\n",
      "rm ontology_data_owl-properties.txt ontology_data_owl-min.owl ontology_data_owl-prefixes.yaml\n",
      "\n",
      "\n",
      "Errors:\n",
      "Picked up _JAVA_OPTIONS: -Xmx4000G\n",
      "relation-graph is running with Java arguments: -Xmx4000G\n",
      "Picked up _JAVA_OPTIONS: -Xmx4000G\n",
      "Error: Could not find or load main class org.renci.relationgraph.Main\n",
      "Caused by: java.lang.ClassNotFoundException: org.renci.relationgraph.Main\n",
      "make: *** [/home/jplfaria/scratch/install_stuff/semsql/semsql/builder/build.Makefile:72: ontology_data_owl-relation-graph.tsv] Error 1\n",
      "\n",
      "CPU times: user 30.1 ms, sys: 34.9 ms, total: 65 ms\n",
      "Wall time: 31min 4s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the database creation\n",
    "from create_semantic_sql_db import create_semantic_sql_db\n",
    "\n",
    "# Run the database creation with custom input filename\n",
    "create_semantic_sql_db(\n",
    "    repo_path,\n",
    "    input_owl_filename='ontology_data_owl.owl'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a59e2574-b743-486e-9962-e6f04a591be5",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Extract SQLite tables to .tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e7e717d-54f1-461b-8a12-7d7cbdfdf3c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading database from: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/CDM_merged_ontologies.db\n",
      "Saving TSV files to: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables\n",
      "Processing table: term_association\n",
      "Exported 'term_association' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/term_association.tsv'\n",
      "Processing table: has_oio_synonym_statement\n",
      "Exported 'has_oio_synonym_statement' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/has_oio_synonym_statement.tsv'\n",
      "Processing table: anonymous_expression\n",
      "Exported 'anonymous_expression' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/anonymous_expression.tsv'\n",
      "Processing table: anonymous_class_expression\n",
      "Exported 'anonymous_class_expression' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/anonymous_class_expression.tsv'\n",
      "Processing table: anonymous_property_expression\n",
      "Exported 'anonymous_property_expression' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/anonymous_property_expression.tsv'\n",
      "Processing table: anonymous_individual_expression\n",
      "Exported 'anonymous_individual_expression' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/anonymous_individual_expression.tsv'\n",
      "Processing table: owl_restriction\n",
      "Exported 'owl_restriction' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/owl_restriction.tsv'\n",
      "Processing table: owl_complex_axiom\n",
      "Exported 'owl_complex_axiom' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/owl_complex_axiom.tsv'\n",
      "Processing table: prefix\n",
      "Exported 'prefix' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/prefix.tsv'\n",
      "Processing table: rdf_list_statement\n",
      "Exported 'rdf_list_statement' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/rdf_list_statement.tsv'\n",
      "Processing table: rdf_level_summary_statistic\n",
      "Exported 'rdf_level_summary_statistic' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/rdf_level_summary_statistic.tsv'\n",
      "Processing table: relation_graph_construct\n",
      "Exported 'relation_graph_construct' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/relation_graph_construct.tsv'\n",
      "Processing table: subgraph_query\n",
      "Exported 'subgraph_query' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/subgraph_query.tsv'\n",
      "Processing table: entailed_edge\n",
      "Exported 'entailed_edge' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/entailed_edge.tsv'\n",
      "Processing table: repair_action\n",
      "Exported 'repair_action' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/repair_action.tsv'\n",
      "Processing table: problem\n",
      "Exported 'problem' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/problem.tsv'\n",
      "Processing table: lexical_problem\n",
      "Exported 'lexical_problem' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/lexical_problem.tsv'\n",
      "Processing table: statements\n",
      "Exported 'statements' to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables/statements.tsv'\n",
      "\n",
      "All tables have been exported to TSV files in: /cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables\n",
      "CPU times: user 24min 6s, sys: 4min 6s, total: 28min 13s\n",
      "Wall time: 28min 13s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Add the scripts directory to the Python path\n",
    "repo_path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "scripts_path = os.path.join(repo_path, 'scripts')\n",
    "sys.path.append(scripts_path)\n",
    "\n",
    "# Import and run the extraction\n",
    "from extract_sql_tables_to_tsv import extract_sql_tables_to_tsv\n",
    "\n",
    "# Run the extraction\n",
    "extract_sql_tables_to_tsv(repo_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "76cc4dce-7860-48cb-a46a-adfe3e052d83",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 'prefix' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/prefix.tsv'\n",
      "Table 'rdf_list_statement' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/rdf_list_statement.tsv'\n",
      "Table 'rdf_level_summary_statistic' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/rdf_level_summary_statistic.tsv'\n",
      "Table 'anonymous_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/anonymous_expression.tsv'\n",
      "Table 'anonymous_class_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/anonymous_class_expression.tsv'\n",
      "Table 'anonymous_property_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/anonymous_property_expression.tsv'\n",
      "Table 'anonymous_individual_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/anonymous_individual_expression.tsv'\n",
      "Table 'owl_restriction' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/owl_restriction.tsv'\n",
      "Table 'owl_complex_axiom' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/owl_complex_axiom.tsv'\n",
      "Table 'has_oio_synonym_statement' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/has_oio_synonym_statement.tsv'\n",
      "Table 'repair_action' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/repair_action.tsv'\n",
      "Table 'problem' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/problem.tsv'\n",
      "Table 'lexical_problem' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/lexical_problem.tsv'\n",
      "Table 'relation_graph_construct' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/relation_graph_construct.tsv'\n",
      "Table 'subgraph_query' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/subgraph_query.tsv'\n",
      "Table 'entailed_edge' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/entailed_edge.tsv'\n",
      "Table 'term_association' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/term_association.tsv'\n",
      "Table 'statements' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/statements.tsv'\n",
      "All tables have been exported to TSV files in the specified directory.\n",
      "CPU times: user 11min 11s, sys: 1min 24s, total: 12min 36s\n",
      "Wall time: 12min 46s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import os\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "#Specify the output directory\n",
    "output_dir = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31/'\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Connect to the SQLite database\n",
    "conn = sqlite3.connect('/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl_31.db')\n",
    "\n",
    "# Get a list of all tables in the database\n",
    "cursor = conn.cursor()\n",
    "cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "tables = cursor.fetchall()\n",
    "\n",
    "# Extract each table and save as TSV\n",
    "for table in tables:\n",
    "    table_name = table[0]\n",
    "    df = pd.read_sql_query(f\"SELECT * FROM {table_name}\", conn)\n",
    "    output_path = os.path.join(output_dir, f\"{table_name}.tsv\")\n",
    "    df.to_csv(output_path, sep='\\t', index=False)\n",
    "    print(f\"Table '{table_name}' has been exported to '{output_path}'\")\n",
    "\n",
    "# Close the connection\n",
    "conn.close()\n",
    "print(\"All tables have been exported to TSV files in the specified directory.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6bc37b06-a091-44ef-95d2-09cd307aa80c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 'term_association' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/term_association.parquet'\n",
      "Table 'has_oio_synonym_statement' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/has_oio_synonym_statement.parquet'\n",
      "Table 'anonymous_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/anonymous_expression.parquet'\n",
      "Table 'anonymous_class_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/anonymous_class_expression.parquet'\n",
      "Table 'anonymous_property_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/anonymous_property_expression.parquet'\n",
      "Table 'anonymous_individual_expression' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/anonymous_individual_expression.parquet'\n",
      "Table 'owl_restriction' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/owl_restriction.parquet'\n",
      "Table 'owl_complex_axiom' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/owl_complex_axiom.parquet'\n",
      "Table 'prefix' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/prefix.parquet'\n",
      "Table 'rdf_list_statement' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/rdf_list_statement.parquet'\n",
      "Table 'rdf_level_summary_statistic' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/rdf_level_summary_statistic.parquet'\n",
      "Table 'relation_graph_construct' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/relation_graph_construct.parquet'\n",
      "Table 'subgraph_query' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/subgraph_query.parquet'\n",
      "Table 'entailed_edge' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/entailed_edge.parquet'\n",
      "Table 'repair_action' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/repair_action.parquet'\n",
      "Table 'problem' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/problem.parquet'\n",
      "Table 'lexical_problem' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/lexical_problem.parquet'\n",
      "Table 'statements' has been exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/statements.parquet'\n",
      "All tables have been exported to Parquet files in the specified directory.\n",
      "CPU times: user 7min 5s, sys: 1min 42s, total: 8min 47s\n",
      "Wall time: 8min 54s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import os\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "\n",
    "# Specify the output directory\n",
    "output_dir = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl'\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Connect to the SQLite database\n",
    "conn = sqlite3.connect('/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl.db')\n",
    "\n",
    "# Get a list of all tables in the database\n",
    "cursor = conn.cursor()\n",
    "cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "tables = cursor.fetchall()\n",
    "\n",
    "# Extract each table and save as Parquet\n",
    "for table in tables:\n",
    "    table_name = table[0]\n",
    "    df = pd.read_sql_query(f\"SELECT * FROM {table_name}\", conn)\n",
    "    output_path = os.path.join(output_dir, f\"{table_name}.parquet\")\n",
    "    df.to_parquet(output_path, index=False)\n",
    "    print(f\"Table '{table_name}' has been exported to '{output_path}'\")\n",
    "\n",
    "# Close the connection\n",
    "conn.close()\n",
    "\n",
    "print(\"All tables have been exported to Parquet files in the specified directory.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a127c894-f60b-4c99-b0a5-51428d375b7b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 200 rows:\n",
      "           stanza       subject         predicate        object  \\\n",
      "0    obo:envo.owl  obo:envo.owl     foaf:homepage          None   \n",
      "1    obo:envo.owl  obo:envo.owl   owl:versionInfo          None   \n",
      "2    obo:envo.owl  obo:envo.owl      rdfs:comment          None   \n",
      "3    obo:envo.owl  obo:envo.owl      rdfs:comment          None   \n",
      "4    obo:envo.owl  obo:envo.owl      rdfs:comment          None   \n",
      "..            ...           ...               ...           ...   \n",
      "195   IAO:0000425   IAO:0000425        rdfs:label          None   \n",
      "196   IAO:0000425   IAO:0000425        rdfs:label          None   \n",
      "197   IAO:0000425   IAO:0000425  rdfs:isDefinedBy  obo:envo.owl   \n",
      "198   IAO:0000425   IAO:0000425       IAO:0000117          None   \n",
      "199   IAO:0000425   IAO:0000425       IAO:0000115          None   \n",
      "\n",
      "                                                 value    datatype language  \\\n",
      "0                      http://environmentontology.org/  xsd:anyURI     None   \n",
      "1                                           2024-07-01        None     None   \n",
      "2    Includes Ontology(OntologyID(OntologyIRI(<http...        None     None   \n",
      "3    Includes Ontology(OntologyID(OntologyIRI(<http...        None     None   \n",
      "4    Includes Ontology(OntologyID(OntologyIRI(<http...        None     None   \n",
      "..                                                 ...         ...      ...   \n",
      "195                                expand assertion to        None       en   \n",
      "196                                expand assertion to        None     None   \n",
      "197                                               None        None     None   \n",
      "198                                      Chris Mungall        None       en   \n",
      "199  A macro expansion tag applied to an annotation...        None       en   \n",
      "\n",
      "    graph  \n",
      "0    None  \n",
      "1    None  \n",
      "2    None  \n",
      "3    None  \n",
      "4    None  \n",
      "..    ...  \n",
      "195  None  \n",
      "196  None  \n",
      "197  None  \n",
      "198  None  \n",
      "199  None  \n",
      "\n",
      "[200 rows x 8 columns]\n",
      "\n",
      "Dataframe Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 37558125 entries, 0 to 37558124\n",
      "Data columns (total 8 columns):\n",
      " #   Column     Dtype \n",
      "---  ------     ----- \n",
      " 0   stanza     object\n",
      " 1   subject    object\n",
      " 2   predicate  object\n",
      " 3   object     object\n",
      " 4   value      object\n",
      " 5   datatype   object\n",
      " 6   language   object\n",
      " 7   graph      object\n",
      "dtypes: object(8)\n",
      "memory usage: 2.2+ GB\n",
      "None\n",
      "CPU times: user 46 s, sys: 29.5 s, total: 1min 15s\n",
      "Wall time: 55.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import pandas as pd\n",
    "\n",
    "# Read the parquet file and display first 200 rows\n",
    "df = pd.read_parquet('/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/statements.parquet')\n",
    "\n",
    "# Display first 200 rows\n",
    "print(\"First 200 rows:\")\n",
    "print(df.head(200))\n",
    "\n",
    "# Also display basic information about the dataframe\n",
    "print(\"\\nDataframe Info:\")\n",
    "print(df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42c2521f-63b6-4b2c-8e07-6f95909925f3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Read the parquet file\n",
    "df = pd.read_parquet('/scratch/jplfaria/KBase_CDM_Ontologies/outputs/parquet_tables_ontology_data_owl/statements.parquet')\n",
    "\n",
    "def extract_prefix(value):\n",
    "    if pd.isna(value):\n",
    "        return None\n",
    "    # Match patterns like 'obo:', 'rdfs:', 'owl:', etc.\n",
    "    match = re.match(r'^([a-zA-Z]+:)', str(value))\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    return None\n",
    "\n",
    "# Extract unique prefixes from subject, predicate, and object columns\n",
    "subject_prefixes = set(df['subject'].apply(extract_prefix).dropna())\n",
    "predicate_prefixes = set(df['predicate'].apply(extract_prefix).dropna())\n",
    "object_prefixes = set(df['object'].apply(extract_prefix).dropna())\n",
    "\n",
    "# Combine all unique prefixes\n",
    "all_prefixes = sorted(subject_prefixes | predicate_prefixes | object_prefixes)\n",
    "\n",
    "print(\"Unique prefixes found in the data:\")\n",
    "for prefix in all_prefixes:\n",
    "    print(f\"- {prefix}\")\n",
    "\n",
    "print(\"\\nTotal number of unique prefixes:\", len(all_prefixes))\n",
    "\n",
    "# Show count of each prefix usage\n",
    "print(\"\\nPrefix usage counts:\")\n",
    "for column in ['subject', 'predicate', 'object']:\n",
    "    prefix_counts = df[column].apply(extract_prefix).value_counts()\n",
    "    print(f\"\\nTop 10 prefixes in {column} column:\")\n",
    "    print(prefix_counts.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "233ddbee-cc22-4a6b-893b-b986c3d5d05e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 'term_association.tsv' to 'term_association.parquet'\n",
      "Converted 'has_oio_synonym_statement.tsv' to 'has_oio_synonym_statement.parquet'\n",
      "Converted 'anonymous_expression.tsv' to 'anonymous_expression.parquet'\n",
      "Converted 'anonymous_class_expression.tsv' to 'anonymous_class_expression.parquet'\n",
      "Converted 'anonymous_property_expression.tsv' to 'anonymous_property_expression.parquet'\n",
      "Converted 'anonymous_individual_expression.tsv' to 'anonymous_individual_expression.parquet'\n",
      "Converted 'owl_restriction.tsv' to 'owl_restriction.parquet'\n",
      "Converted 'owl_complex_axiom.tsv' to 'owl_complex_axiom.parquet'\n",
      "Converted 'prefix.tsv' to 'prefix.parquet'\n",
      "Converted 'rdf_list_statement.tsv' to 'rdf_list_statement.parquet'\n",
      "Converted 'rdf_level_summary_statistic.tsv' to 'rdf_level_summary_statistic.parquet'\n",
      "Converted 'relation_graph_construct.tsv' to 'relation_graph_construct.parquet'\n",
      "Converted 'subgraph_query.tsv' to 'subgraph_query.parquet'\n",
      "Converted 'entailed_edge.tsv' to 'entailed_edge.parquet'\n",
      "Converted 'repair_action.tsv' to 'repair_action.parquet'\n",
      "Converted 'problem.tsv' to 'problem.parquet'\n",
      "Converted 'lexical_problem.tsv' to 'lexical_problem.parquet'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<timed exec>:23: DtypeWarning: Columns (4,5,6) have mixed types. Specify dtype option on import or set low_memory=False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 'statements.tsv' to 'statements.parquet'\n",
      "\n",
      "Conversion complete! All TSV files have been converted to Parquet format.\n",
      "CPU times: user 4min 31s, sys: 50.7 s, total: 5min 21s\n",
      "Wall time: 5min 59s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Input and output directories\n",
    "input_dir = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl'\n",
    "output_dir = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/parquet_tables_ontology_data_owl'\n",
    "\n",
    "# Ensure output directory exists\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Get all TSV files in the input directory\n",
    "tsv_files = [f for f in os.listdir(input_dir) if f.endswith('.tsv')]\n",
    "\n",
    "# Convert each TSV file to Parquet\n",
    "for tsv_file in tsv_files:\n",
    "    # Construct full file paths\n",
    "    tsv_path = os.path.join(input_dir, tsv_file)\n",
    "    parquet_file = tsv_file.replace('.tsv', '.parquet')\n",
    "    parquet_path = os.path.join(output_dir, parquet_file)\n",
    "    \n",
    "    # Read TSV and write to Parquet\n",
    "    try:\n",
    "        df = pd.read_csv(tsv_path, sep='\\t')\n",
    "        df.to_parquet(parquet_path, index=False)\n",
    "        print(f\"Converted '{tsv_file}' to '{parquet_file}'\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error converting {tsv_file}: {str(e)}\")\n",
    "\n",
    "print(\"\\nConversion complete! All TSV files have been converted to Parquet format.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "94bfd3a5-7ce9-4772-8d3c-2188de0679d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 'term_association' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/term_association.tsv'\n",
      "Table 'has_oio_synonym_statement' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/has_oio_synonym_statement.tsv'\n",
      "Table 'anonymous_expression' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/anonymous_expression.tsv'\n",
      "Table 'anonymous_class_expression' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/anonymous_class_expression.tsv'\n",
      "Table 'anonymous_property_expression' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/anonymous_property_expression.tsv'\n",
      "Table 'anonymous_individual_expression' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/anonymous_individual_expression.tsv'\n",
      "Table 'owl_restriction' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/owl_restriction.tsv'\n",
      "Table 'owl_complex_axiom' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/owl_complex_axiom.tsv'\n",
      "Table 'prefix' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/prefix.tsv'\n",
      "Table 'rdf_list_statement' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/rdf_list_statement.tsv'\n",
      "Table 'rdf_level_summary_statistic' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/rdf_level_summary_statistic.tsv'\n",
      "Table 'relation_graph_construct' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/relation_graph_construct.tsv'\n",
      "Table 'subgraph_query' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/subgraph_query.tsv'\n",
      "Table 'entailed_edge' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/entailed_edge.tsv'\n",
      "Table 'repair_action' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/repair_action.tsv'\n",
      "Table 'problem' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/problem.tsv'\n",
      "Table 'lexical_problem' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/lexical_problem.tsv'\n",
      "Table 'statements' has been exported to '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec/statements.tsv'\n",
      "All tables have been exported to TSV files in the specified directory.\n",
      "CPU times: user 1min 50s, sys: 0 ns, total: 1min 50s\n",
      "Wall time: 1min 49s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import os\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "#Specify the output directory\n",
    "output_dir = '/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl_ec'\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Connect to the SQLite database\n",
    "conn = sqlite3.connect('/cdm_shared_workspace/user_shared_workspace/Jose/KBase_CDM_Ontologies/outputs/eccode.db')\n",
    "\n",
    "# Get a list of all tables in the database\n",
    "cursor = conn.cursor()\n",
    "cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "tables = cursor.fetchall()\n",
    "\n",
    "# Extract each table and save as TSV\n",
    "for table in tables:\n",
    "    table_name = table[0]\n",
    "    df = pd.read_sql_query(f\"SELECT * FROM {table_name}\", conn)\n",
    "    output_path = os.path.join(output_dir, f\"{table_name}.tsv\")\n",
    "    df.to_csv(output_path, sep='\\t', index=False)\n",
    "    print(f\"Table '{table_name}' has been exported to '{output_path}'\")\n",
    "\n",
    "# Close the connection\n",
    "conn.close()\n",
    "print(\"All tables have been exported to TSV files in the specified directory.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba1b9a85-0790-4a46-a464-528af5eeec56",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 'term_association' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/term_association_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'has_oio_synonym_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_oio_synonym_statement_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'anonymous_expression' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_expression_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'anonymous_class_expression' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_class_expression_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'anonymous_property_expression' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_property_expression_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'anonymous_individual_expression' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_individual_expression_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'owl_restriction' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_restriction_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'owl_complex_axiom' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_complex_axiom_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'prefix' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/prefix_table.tsv'\n",
      "Number of rows: 1120\n",
      "Table 'rdf_list_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/rdf_list_statement_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'rdf_level_summary_statistic' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/rdf_level_summary_statistic_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'relation_graph_construct' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/relation_graph_construct_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'subgraph_query' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/subgraph_query_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'entailed_edge' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/entailed_edge_table.tsv'\n",
      "Number of rows: 117371459\n",
      "Table 'repair_action' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/repair_action_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'problem' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/problem_table.tsv'\n",
      "Number of rows: 0\n",
      "Table 'lexical_problem' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/lexical_problem_table.tsv'\n",
      "Number of rows: 0\n",
      "View 'has_text_definition_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_text_definition_statement_view.tsv'\n",
      "Number of rows: 318857\n",
      "View 'has_exact_synonym_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_exact_synonym_statement_view.tsv'\n",
      "Number of rows: 373387\n",
      "View 'has_broad_synonym_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_broad_synonym_statement_view.tsv'\n",
      "Number of rows: 9499\n",
      "View 'has_narrow_synonym_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_narrow_synonym_statement_view.tsv'\n",
      "Number of rows: 22726\n",
      "View 'has_related_synonym_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_related_synonym_statement_view.tsv'\n",
      "Number of rows: 567139\n",
      "View 'has_synonym_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_synonym_statement_view.tsv'\n",
      "Number of rows: 972750\n",
      "View 'has_exact_match_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_exact_match_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'has_broad_match_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_broad_match_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'has_narrow_match_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_narrow_match_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'has_related_match_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_related_match_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'has_match_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_match_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'has_dbxref_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_dbxref_statement_view.tsv'\n",
      "Number of rows: 4072750\n",
      "View 'has_mapping_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_mapping_statement_view.tsv'\n",
      "Number of rows: 4072750\n",
      "View 'contributor' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/contributor_view.tsv'\n",
      "Number of rows: 13931\n",
      "View 'creator' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/creator_view.tsv'\n",
      "Number of rows: 98\n",
      "View 'orcid' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/orcid_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'axiom_dbxref_annotation' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/axiom_dbxref_annotation_view.tsv'\n",
      "Number of rows: 636653\n",
      "View 'ontology_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/ontology_node_view.tsv'\n",
      "Number of rows: 1\n",
      "View 'object_property_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/object_property_node_view.tsv'\n",
      "Number of rows: 839\n",
      "View 'transitive_property_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/transitive_property_node_view.tsv'\n",
      "Number of rows: 56\n",
      "View 'symmetric_property_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/symmetric_property_node_view.tsv'\n",
      "Number of rows: 89\n",
      "View 'reflexive_property_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/reflexive_property_node_view.tsv'\n",
      "Number of rows: 2\n",
      "View 'irreflexive_property_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/irreflexive_property_node_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'asymmetric_property_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/asymmetric_property_node_view.tsv'\n",
      "Number of rows: 2\n",
      "View 'annotation_property_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/annotation_property_node_view.tsv'\n",
      "Number of rows: 432\n",
      "View 'deprecated_node' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/deprecated_node_view.tsv'\n",
      "Number of rows: 37097\n",
      "View 'owl_imports_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_imports_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'owl_inverse_of_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_inverse_of_statement_view.tsv'\n",
      "Number of rows: 139\n",
      "View 'owl_complement_of_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_complement_of_statement_view.tsv'\n",
      "Number of rows: 238\n",
      "View 'owl_equivalent_class_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_equivalent_class_statement_view.tsv'\n",
      "Number of rows: 25229\n",
      "View 'owl_same_as_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_same_as_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'owl_disjoint_class_statement' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_disjoint_class_statement_view.tsv'\n",
      "Number of rows: 0\n",
      "View 'owl_reified_axiom' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_reified_axiom_view.tsv'\n",
      "Number of rows: 1190629\n",
      "Error processing view owl_axiom: Execution failed on sql 'SELECT * FROM owl_axiom': SELECTs to the left and right of UNION do not have the same number of result columns\n",
      "View 'owl_axiom_annotation' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_axiom_annotation_view.tsv'\n",
      "Number of rows: 1394483\n",
      "View 'owl_some_values_from' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_some_values_from_view.tsv'\n",
      "Number of rows: 1670200\n",
      "View 'owl_all_values_from' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_all_values_from_view.tsv'\n",
      "Number of rows: 328\n",
      "View 'owl_has_value' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_has_value_view.tsv'\n",
      "Number of rows: 134\n",
      "View 'owl_has_self' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_has_self_view.tsv'\n",
      "Number of rows: 1\n",
      "View 'owl_subclass_of_some_values_from' exported to '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_subclass_of_some_values_from_view.tsv'\n",
      "Number of rows: 1635417\n",
      "Error processing view owl_equivalent_to_intersection_member: Execution failed on sql 'SELECT * FROM owl_equivalent_to_intersection_member': database or disk is full\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import os\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "\n",
    "def export_database_objects(db_path, output_dir):\n",
    "    # Ensure the output directory exists\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    # Connect to the SQLite database\n",
    "    conn = sqlite3.connect(db_path)\n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    # Get both tables and views\n",
    "    cursor.execute(\"\"\"\n",
    "        SELECT name, type \n",
    "        FROM sqlite_master \n",
    "        WHERE type IN ('table', 'view')\n",
    "        AND name NOT LIKE 'sqlite_%';\n",
    "    \"\"\")\n",
    "    \n",
    "    db_objects = cursor.fetchall()\n",
    "    \n",
    "    # Extract each object and save as TSV\n",
    "    for obj_name, obj_type in db_objects:\n",
    "        try:\n",
    "            df = pd.read_sql_query(f\"SELECT * FROM {obj_name}\", conn)\n",
    "            output_path = os.path.join(output_dir, f\"{obj_name}_{obj_type}.tsv\")\n",
    "            df.to_csv(output_path, sep='\\t', index=False)\n",
    "            print(f\"{obj_type.capitalize()} '{obj_name}' exported to '{output_path}'\")\n",
    "            print(f\"Number of rows: {len(df)}\")\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing {obj_type} {obj_name}: {str(e)}\")\n",
    "    \n",
    "    conn.close()\n",
    "    print(\"\\nExport complete!\")\n",
    "\n",
    "# Usage\n",
    "db_path = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl.db'\n",
    "output_dir = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl'\n",
    "\n",
    "export_database_objects(db_path, output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ac26be3-cee9-4660-bc77-8fff04bd16b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================================================================================\n",
      "Table: 'term_association'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- evidence_type\n",
      "- publication\n",
      "- source\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/term_association_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'has_oio_synonym_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_oio_synonym_statement_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'anonymous_expression'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_expression_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'anonymous_class_expression'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_class_expression_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'anonymous_property_expression'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_property_expression_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'anonymous_individual_expression'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/anonymous_individual_expression_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'owl_restriction'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- on_property\n",
      "- filler\n",
      "- id\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_restriction_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'owl_complex_axiom'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_complex_axiom_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'prefix'\n",
      "Number of rows: 1120\n",
      "\n",
      "Columns:\n",
      "- prefix\n",
      "- base\n",
      "\n",
      "First row of data:\n",
      "  prefix: prefix\n",
      "  base: base\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/prefix_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'rdf_list_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/rdf_list_statement_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'rdf_level_summary_statistic'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- element\n",
      "- count_value\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/rdf_level_summary_statistic_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'relation_graph_construct'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/relation_graph_construct_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'subgraph_query'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- anchor_object\n",
      "- anchor_predicate\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/subgraph_query_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'entailed_edge'\n",
      "Number of rows: 117371459\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "\n",
      "First row of data:\n",
      "  subject: CHEBI:165710\n",
      "  predicate: rdfs:subClassOf\n",
      "  object: CHEBI:165710\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/entailed_edge_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'repair_action'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- description\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/repair_action_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'problem'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- value\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/problem_table.tsv\n",
      "\n",
      "================================================================================\n",
      "Table: 'lexical_problem'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- subject\n",
      "- predicate\n",
      "- value\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/lexical_problem_table.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_text_definition_statement'\n",
      "Number of rows: 318857\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: BFO:0000179\n",
      "  subject: BFO:0000179\n",
      "  predicate: IAO:0000115\n",
      "  object: None\n",
      "  value: Relates an entity in the ontology to the name...\n",
      "  datatype: None\n",
      "  language: en\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_text_definition_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_exact_synonym_statement'\n",
      "Number of rows: 373387\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: OMO:0003010\n",
      "  subject: OMO:0003010\n",
      "  predicate: oio:hasExactSynonym\n",
      "  object: None\n",
      "  value: INN\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_exact_synonym_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_broad_synonym_statement'\n",
      "Number of rows: 9499\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: RO:0002374\n",
      "  subject: RO:0002374\n",
      "  predicate: oio:hasBroadSynonym\n",
      "  object: None\n",
      "  value: has fused element\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_broad_synonym_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_narrow_synonym_statement'\n",
      "Number of rows: 22726\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: RO:0002553\n",
      "  subject: RO:0002553\n",
      "  predicate: oio:hasNarrowSynonym\n",
      "  object: None\n",
      "  value: hyperparasitoid of\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_narrow_synonym_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_related_synonym_statement'\n",
      "Number of rows: 567139\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: RO:0002248\n",
      "  subject: RO:0002248\n",
      "  predicate: oio:hasRelatedSynonym\n",
      "  object: None\n",
      "  value: has active pharmaceutical ingredient\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_related_synonym_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_synonym_statement'\n",
      "Number of rows: 972750\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: <https://bioregistry.io/eccode:1.1.1.101>\n",
      "  subject: <https://bioregistry.io/eccode:1.1.1.101>\n",
      "  predicate: oio:hasExactSynonym\n",
      "  object: None\n",
      "  value: palmitoyldihydroxyacetone-phosphate reductase\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_synonym_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_exact_match_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_exact_match_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_broad_match_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_broad_match_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_narrow_match_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_narrow_match_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_related_match_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_related_match_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_match_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_match_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_dbxref_statement'\n",
      "Number of rows: 4072750\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: IAO:0000233\n",
      "  subject: IAO:0000233\n",
      "  predicate: oio:hasDbXref\n",
      "  object: None\n",
      "  value: IAO:0000233\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_dbxref_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'has_mapping_statement'\n",
      "Number of rows: 4072750\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: <https://purl.obolibrary.org/obo/PCO_0000040>\n",
      "  subject: <https://purl.obolibrary.org/obo/PCO_0000040>\n",
      "  predicate: oio:hasDbXref\n",
      "  object: None\n",
      "  value: ...\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/has_mapping_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'contributor'\n",
      "Number of rows: 13931\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: obo:foodon.owl\n",
      "  subject: obo:foodon.owl\n",
      "  predicate: dcterms:contributor\n",
      "  object: None\n",
      "  value: Rhiannon Cameron\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/contributor_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'creator'\n",
      "Number of rows: 98\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: RO:0002575\n",
      "  subject: RO:0002575\n",
      "  predicate: dcterms:creator\n",
      "  object: orcid:0000-0002-6601-2165\n",
      "  value: None\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/creator_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'orcid'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "- label\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/orcid_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'axiom_dbxref_annotation'\n",
      "Number of rows: 636653\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- id\n",
      "- annotation_subject\n",
      "- annotation_predicate\n",
      "- annotation_object\n",
      "- annotation_value\n",
      "- annotation_language\n",
      "- annotation_datatype\n",
      "\n",
      "First row of data:\n",
      "  stanza: obo:UBPROP_0000001\n",
      "  subject: obo:UBPROP_0000001\n",
      "  predicate: IAO:0000115\n",
      "  object: None\n",
      "  value: An alternate textual definition for a class...\n",
      "  datatype: None\n",
      "  language: None\n",
      "  id: _:riog00000125\n",
      "  annotation_subject: _:riog00000125\n",
      "  annotation_predicate: oio:hasDbXref\n",
      "  annotation_object: None\n",
      "  annotation_value: https://orcid.org/0000-0002-6601-2165\n",
      "  annotation_language: None\n",
      "  annotation_datatype: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/axiom_dbxref_annotation_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'ontology_node'\n",
      "Number of rows: 1\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: obo:foodon.owl\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/ontology_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'object_property_node'\n",
      "Number of rows: 839\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: BFO:0000050\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/object_property_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'transitive_property_node'\n",
      "Number of rows: 56\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: BFO:0000050\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/transitive_property_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'symmetric_property_node'\n",
      "Number of rows: 89\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: FOODON:00001301\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/symmetric_property_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'reflexive_property_node'\n",
      "Number of rows: 2\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: RO:0002351\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/reflexive_property_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'irreflexive_property_node'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/irreflexive_property_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'asymmetric_property_node'\n",
      "Number of rows: 2\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: RO:0002551\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/asymmetric_property_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'annotation_property_node'\n",
      "Number of rows: 432\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: <http://geneontology.org/foo/applies-pattern>\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/annotation_property_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'deprecated_node'\n",
      "Number of rows: 37097\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "\n",
      "First row of data:\n",
      "  id: BFO:0000060\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/deprecated_node_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_imports_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_imports_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_inverse_of_statement'\n",
      "Number of rows: 139\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: BFO:0000050\n",
      "  subject: BFO:0000050\n",
      "  predicate: owl:inverseOf\n",
      "  object: BFO:0000051\n",
      "  value: None\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_inverse_of_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_complement_of_statement'\n",
      "Number of rows: 238\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: IAO:0000583\n",
      "  subject: _:riog00000573\n",
      "  predicate: owl:complementOf\n",
      "  object: IAO:0000416\n",
      "  value: None\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_complement_of_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_equivalent_class_statement'\n",
      "Number of rows: 25229\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "First row of data:\n",
      "  stanza: BFO:0000023\n",
      "  subject: BFO:0000023\n",
      "  predicate: owl:equivalentClass\n",
      "  object: CHEBI:50906\n",
      "  value: None\n",
      "  datatype: None\n",
      "  language: None\n",
      "  graph: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_equivalent_class_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_same_as_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_same_as_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_disjoint_class_statement'\n",
      "Number of rows: 0\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- graph\n",
      "\n",
      "No data in this table/view\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_disjoint_class_statement_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_reified_axiom'\n",
      "Number of rows: 1190629\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "\n",
      "First row of data:\n",
      "  id: _:riog00000125\n",
      "  stanza: obo:UBPROP_0000001\n",
      "  subject: obo:UBPROP_0000001\n",
      "  predicate: IAO:0000115\n",
      "  object: None\n",
      "  value: An alternate textual definition for a class...\n",
      "  datatype: None\n",
      "  language: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_reified_axiom_view.tsv\n",
      "Error processing view owl_axiom: Execution failed on sql 'SELECT * FROM owl_axiom': SELECTs to the left and right of UNION do not have the same number of result columns\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_axiom_annotation'\n",
      "Number of rows: 1394483\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "- value\n",
      "- datatype\n",
      "- language\n",
      "- id\n",
      "- annotation_subject\n",
      "- annotation_predicate\n",
      "- annotation_object\n",
      "- annotation_value\n",
      "- annotation_language\n",
      "- annotation_datatype\n",
      "\n",
      "First row of data:\n",
      "  stanza: obo:UBPROP_0000001\n",
      "  subject: obo:UBPROP_0000001\n",
      "  predicate: IAO:0000115\n",
      "  object: None\n",
      "  value: An alternate textual definition for a class...\n",
      "  datatype: None\n",
      "  language: None\n",
      "  id: _:riog00000125\n",
      "  annotation_subject: _:riog00000125\n",
      "  annotation_predicate: oio:hasDbXref\n",
      "  annotation_object: None\n",
      "  annotation_value: https://orcid.org/0000-0002-6601-2165\n",
      "  annotation_language: None\n",
      "  annotation_datatype: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_axiom_annotation_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_some_values_from'\n",
      "Number of rows: 1670200\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "- on_property\n",
      "- filler\n",
      "\n",
      "First row of data:\n",
      "  id: _:riog00000637\n",
      "  on_property: RO:0000087\n",
      "  filler: OBI:0000202\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_some_values_from_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_all_values_from'\n",
      "Number of rows: 328\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "- on_property\n",
      "- filler\n",
      "\n",
      "First row of data:\n",
      "  id: _:riog00002313\n",
      "  on_property: BFO:0000050\n",
      "  filler: BFO:0000002\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_all_values_from_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_has_value'\n",
      "Number of rows: 134\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "- on_property\n",
      "- filler\n",
      "\n",
      "First row of data:\n",
      "  id: _:riog04243142\n",
      "  on_property: OBI:0000304\n",
      "  filler: OBI:0000758\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_has_value_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_has_self'\n",
      "Number of rows: 1\n",
      "\n",
      "Columns:\n",
      "- id\n",
      "- on_property\n",
      "- filler\n",
      "\n",
      "First row of data:\n",
      "  id: _:riog01057611\n",
      "  on_property: RO:0002481\n",
      "  filler: None\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_has_self_view.tsv\n",
      "\n",
      "================================================================================\n",
      "View: 'owl_subclass_of_some_values_from'\n",
      "Number of rows: 1635417\n",
      "\n",
      "Columns:\n",
      "- stanza\n",
      "- subject\n",
      "- predicate\n",
      "- object\n",
      "\n",
      "First row of data:\n",
      "  stanza: CARO:0000014\n",
      "  subject: CARO:0000014\n",
      "  predicate: BFO:0000050\n",
      "  object: CARO:0000013\n",
      "\n",
      "Exported to: /scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl/owl_subclass_of_some_values_from_view.tsv\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "import textwrap\n",
    "\n",
    "def export_database_objects(db_path, output_dir):\n",
    "    # Ensure the output directory exists\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    # Connect to the SQLite database\n",
    "    conn = sqlite3.connect(db_path)\n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    # Get both tables and views\n",
    "    cursor.execute(\"\"\"\n",
    "        SELECT name, type \n",
    "        FROM sqlite_master \n",
    "        WHERE type IN ('table', 'view')\n",
    "        AND name NOT LIKE 'sqlite_%';\n",
    "    \"\"\")\n",
    "    \n",
    "    db_objects = cursor.fetchall()\n",
    "    \n",
    "    # Extract each object and save as TSV\n",
    "    for obj_name, obj_type in db_objects:\n",
    "        try:\n",
    "            # Read the data into a DataFrame\n",
    "            df = pd.read_sql_query(f\"SELECT * FROM {obj_name}\", conn)\n",
    "            \n",
    "            # Save to TSV\n",
    "            output_path = os.path.join(output_dir, f\"{obj_name}_{obj_type}.tsv\")\n",
    "            df.to_csv(output_path, sep='\\t', index=False)\n",
    "            \n",
    "            # Print detailed information\n",
    "            print(\"\\n\" + \"=\"*80)\n",
    "            print(f\"{obj_type.capitalize()}: '{obj_name}'\")\n",
    "            print(f\"Number of rows: {len(df)}\")\n",
    "            print(\"\\nColumns:\")\n",
    "            for col in df.columns:\n",
    "                print(f\"- {col}\")\n",
    "                \n",
    "            if len(df) > 0:\n",
    "                print(\"\\nFirst row of data:\")\n",
    "                # Format the first row nicely\n",
    "                first_row = df.iloc[0]\n",
    "                for col, val in first_row.items():\n",
    "                    # Wrap long values for better readability\n",
    "                    if isinstance(val, str) and len(str(val)) > 50:\n",
    "                        val = textwrap.shorten(str(val), width=50, placeholder=\"...\")\n",
    "                    print(f\"  {col}: {val}\")\n",
    "            else:\n",
    "                print(\"\\nNo data in this table/view\")\n",
    "            \n",
    "            print(f\"\\nExported to: {output_path}\")\n",
    "            \n",
    "        except Exception as e:\n",
    "            print(f\"Error processing {obj_type} {obj_name}: {str(e)}\")\n",
    "    \n",
    "    conn.close()\n",
    "    print(\"\\nExport complete!\")\n",
    "\n",
    "# Usage\n",
    "db_path = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/ontology_data_owl.db'\n",
    "output_dir = '/scratch/jplfaria/KBase_CDM_Ontologies/outputs/tsv_tables_ontology_data_owl'\n",
    "\n",
    "export_database_objects(db_path, output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2bba10c-a198-4b79-964d-7c0895213f7e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3_ModelSEED",
   "language": "python",
   "name": "python3_modelseed"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
